* RNA-seq Snakemake Pipelines
:PROPERTIES:
:ID:       2a6a5dac-151e-42e8-a80a-74f77ad0d4ca
:header-args: :tangle no :tangle-mode (identity #o555) :mkdirp yes
:header-args+: :noweb yes
:END:
** Processes fro analysis of gene expression
:PROPERTIES:
:ID:       4c2a51ad-930a-40b9-b074-191a00b3110d
:END:
*** Bulk RNA-seq
:PROPERTIES:
:ID:       4206be38-8d3c-43c4-a5c7-ca6a8bd8361a
:CREATED:  [2021-02-17 Wed 09:08]
:END:
- http://bioinformatics.sdstate.edu/idep/
- https://nf-co.re/rnaseq/3.13.2
- https://www.nature.com/articles/s41598-020-76881-x
- https://www.cell.com/ajhg/fulltext/S0002-9297(18)30401-4
**** RNA-seq experimental design
:PROPERTIES:
:ID:       19780d7c-afa7-4195-b1f9-df9c563de303
:END:
- https://www.biostars.org/p/9549394/
- https://hbctraining.github.io/DGE_workshop/lessons/08_DGE_LRT.html
- https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7873980/
- https://bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/designmatrices.html#design-matrices-with-and-without-intercept-term
- https://bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/designmatrices.html#design-matrices-with-and-without-intercept-term
- https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2021/RNAseq/Markdowns/09_Linear_Models.html
- spike-ins
  - https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4538800/
- validation technologies
  - cannot distinguish isomiRs
  - qPCR expected to correlate >.9 w/ seq expression > 2^8 cite:yeri2018
  - qPCR more concordant than EdgeSeq or FirePlex cite:yeri2018
- what depth? https://www.researchgate.net/post/how_many_million_reads_for_miRNA_discovery
- reads per sample
  - https://www.qiagen.com/us/resources/faq?id=a119d5f0-0a2a-4612-828e-d8fa3cc37891&lang=en
- cite:garve2012
- cite:williams2014
- https://www.degruyter.com/view/journals/sagmb/18/1/article-20180021.xml?language=en
- https://ucdavis-bioinformatics-training.github.io/2018-June-RNA-Seq-Workshop/tuesday/ExperimentalDesign.pdf
- http://scotty.genetics.utah.edu/help.html
- Depth
  - - determine directly using technical replicates
  - expression estimate correspondence among technical replicates (no biological variation) for genes with above-median coverage stabilized at about 10 million reads per sample (Wang et al. 2011).
- Replicates and power
  - sufficient replication has to occur at every level of comparison. In a fully "factorial" design, for example, more than one experimental "factor" is of interest, each with two or more treatments, such that any individual receives one of multiple possible treatments at each factor. In this case, every possible combination of treatments across factors needs to be replicated sufficiently,
  - biological variation is more important than technical variation- See Bullard,10 Evaluation of statistical methods for normalization and differential expression in mRNA-Seq experiments, fig. 4
  - adding depth vs. replicate- replicates preferred, gives less uncertainty
  - remember, more comparisons require more replicates
  - cite:schurch2016
  - Power
    - https://cofactorgenomics.com/6-changes-thatll-make-big-difference-rna-seq-part-5/
    - cite:peixoto2015
    - the rewards in statistical power due to additional biological replicates will surpass the improved parameterization of technical variation garnered from additional technical replicates
    - low counts mean low power- see rna seqlopedia fig 1.1
  - [[http://scotty.genetics.utah.edu/help.html][Scotty]]
  - consider a "bank" of more sample replicates than are acutally sequenced
- By specific experimental question
  - Differential expression
    - [[http://rnaseq.uoregon.edu/#exp-design][RNA-seqlopedia]]
      - qualitative or quantitative (DE, ect.) design- usually quant
      - need bio replicates
      - stranded and paired library prep not required
- Prefer “balanced block” design w/ barcode-> mixed prep-> all samples on all lanes
- cite:auer2010

***** General reference
:PROPERTIES:
:ID:       6a916054-5e00-4833-a44c-e6d815ffd07f
:END:
- http://www.slideshare.net/YaoyuEWang/rnaseq-analysis-ngsapplication2
- http://michelebusby.tumblr.com/post/26913184737/thinking-about-designing-rna-seq-experiments-to
***** Parameters
:PROPERTIES:
:ID:       2edf5837-5b50-490f-9e61-73151a280483
:END:
- cite:williams2014
  - contains the 30-50 mil read for DE recommendation
****** Parameters affecting sequencing
:PROPERTIES:
:ID:       809ff66c-4f53-46f5-b251-95e51baf6bf0
:END:
  - Variance is bimodal
    - low @ low counts b/c Poisson
    - low @ high counts b/c?
  - biological
    - transcript abundence 10^5-10^7 see cite:Griffith_2015 refs 85,86
  - technical @ platform
    - specialized
      - On target efficiency with bedtools
    - read length
    - read depth
    - strandedness problem[[file:org.org_imgs/20170224_132238_567aRU.png]]
  - Strandedness
    - cite:Zhao_2015

***** Validation
:PROPERTIES:
:ID:       0c8f144f-2874-404e-8c0e-160065fbfa94
:END:
- validation technologies
  - cannot distinguish isomiRs
  - qPCR expected to correlate >.9 w/ seq expression > 2^8 cite:yeri2018
  - qPCR more concordant than EdgeSeq or FirePlex cite:yeri2018
**** [[id:386d899a-9208-4d5c-93d8-3750b22d9748][RNA bench methods]]
**** [[id:7BF989EC-4DA7-4B0C-AAA9-0175156447FF][RNA extraction, solubilization, and isolation]]
**** [[id:4E23127E-4ED6-4802-A03C-4103AA8CAF29][RNA Quality and Quantification]]
**** Library preparation
:PROPERTIES:
:ID:       2111e30d-9183-40de-948e-3f76b70905d8
:END:
  - common kits
    - RNA exome
    - pico
  - From input to library
    - cite:van_Dijk_2014
    - By workflow
      - Isolation
      - Enrichment
        - Methods
          - Hybridization
            - For selection of targets
              - via oligo-dT
                - will include some mitochondial
            - For removal of non-targets
              - rRNA depletion- rely on specific oligo set, d/n work well if rRNA is degraded
          - Size-selection enrichment- for ncRNA
          - DSN (duplex-specific nuclease digestion)- @ cDNA level, allow abundant transcripts to renature and are degraded by duplex-specific nuclease
      - Fragmentation- RNase III, metal ion, heat, or sonication
      - Library preparation
        - FFPE
          - However, formalin treatment degrades RNA through cross-linking and fragmentation, which significantly impairs molecular analyses (Bass et al., 2014; Farragher et al., 2008; Klopfleisch et al., 2011). T
          - check webster 15 cites- (Auerbach et al., 2014; Hedegaard et al., 2014; Linton et al., 2012; Spencer et al., 2013; Zhao et al., 2014b).
        - [[http://support.illumina.com/content/dam/illumina-support/courses/truseq-rna-access/story.html?iframe?iframe][TruSeq: RNA Access Library Prep]]
        - [[http://support.illumina.com/content/dam/illumina-support/courses/TruSeq_Sample_Prep_Best_Practices/index.html?iframe?iframe][TruSeq Sample Preparation Best Practices]]
        - First-strand synthesis
          - Priming options
            - oligo-dT
            - "random" hex
        - Second-strand synthesis
          - Also requires a primer
          - Usually via RNA displacement- RNase H and T4 ligase
        - cDNA fragmentation
          - accounting for fragment bias- https://genomebiology.biomedcentral.com/articles/10.1186/gb-2011-12-3-r22
        - Sequencing adaptors
          - Y-adaptor PCR: blunt cDNA ends, add adenosine overhangs
- cite:oneil2013
- stranded
  - https://www.ecseq.com/support/ngs/how-do-strand-specific-sequencing-protocols-work
**** [[id:3BC41E6A-00D3-4498-A48C-463F68CFD76B][Massively parallel sequencing]]
:PROPERTIES:
:ID:       68056b25-1413-4bda-a1f2-6610e92c6765
:END:
- what depth? https://www.researchgate.net/post/how_many_million_reads_for_miRNA_discovery
**** [[id:1EE731A3-83DD-46AF-B988-2311307C335F][Raw sequence quality control]]
**** Alignment and pseudoalignment
:PROPERTIES:
:ID:       5f9ec48c-4be0-4351-abff-348144b96d0f
:END:
- http://refgenomes.databio.org/v3/assets/archive/2230c535660fb4774114bfa966a62f823fdb6d21acf138d4/salmon_partial_sa_index?tag=default

- pseudoaligner
  - kallisto
  - Salmon
- Aligners
  - TopHat2
  - Specialized aligners
    - Fusions
      - TopHat Fusion
        - tophat fusion documentation http://ccb.jhu.edu/software/tophat/fusion_manual.html
      - STAR for chimeric reads / STAR-Fusion
        - [[https://github.com/STAR-Fusion/STAR-Fusion/wiki][website]]
        - From cite:star_2.5_dobin Chapter 5
        - From cite:dobin2015mapping Alternate Protocol 6
          - --chimSegmentMin <N> is minimum allowed length for each of the chimeric segments
        - additional outputs are
          - Chimeric.out.sam
          - Chimeric.out.junction- a STAR-specific format
    - To transcriptome
      -
  - [[*STAR][STAR]]
- HISAT2

***** [[id:326ecd60-8cd4-4815-a389-967b2c3fef0a][Nucleic acid sequence alignment]]
:PROPERTIES:
:ID:       542988fa-307f-454e-927e-ee6a17187265
:END:
***** Pseudoalignment
:PROPERTIES:
:ID:       6518d9c5-4b68-4c7a-873d-2468fd30dfcd
:END:
- https://www.reddit.com/r/bioinformatics/comments/uv1hu9/what_are_the_pros_and_cons_of_pseudo_alignment/
***** Read annotation
:PROPERTIES:
:ID:       8e435b88-6895-4e75-a728-cfdaa272bddb
:END:
- https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-015-1308-8
- https://github.com/griffithlab/rnaseq_tutorial/wiki/Annotation
**** Quantification and normalization
:PROPERTIES:
:ID:       853dd34c-27be-4be8-9794-19f04a1e22b0
:END:
- via spike or replicate with RUVs https://www.nature.com/articles/nbt.2931?error=cookies_not_supported&code=a01fb01e-6702-4e1e-96da-1dbad752e672
- https://www.reneshbedre.com/blog/expression_units.html
- http://luisvalesilva.com/datasimple/rna-seq_units.html
https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html
by spike-in  Jiang,L., Schlesinger,F., Davis,C.A., Zhang,Y., Li,R., Salit,M., Gingeras,T.R. and Oliver,B. (2011) Synthetic spike-in standards for RNA-seq experiments. Genome Res., 21, 1543–1551.

- choosing a strategy- Zyprych-Walczak,J., Szabelska,A., Handschuh,L., Górczak,K., Klamecka,K., Figlerowicz,M. and Siatkowski,I. (2015) The impactof normalization methods on RNA-Seq data analysis. Biomed Res. Int., 2015, doi:10.1155/2015/621690.
- what differences are controlled?
  - library size
  - batch effect
  - gene length
  - GC content
- methods
  - upper quartile- remove zero counts, scale based on 75th quartile of remaining reads Bullard,J.H., Purdom,E., Hansen,K.D. and Dudoit,S. (2010) Evaluation of statistical methods for normalization and differential expression in mRNA-Seq experiments. BMC Bioinformatics, 11, doi:10.1186/1471-2105-11-94.
  - quantile- read distribution assumed identical, counts adjusted by median of each quantile across all samples
  - TMM
  - pseudoalignment
    - kallisto, used with sleuth for DE
      - https://www.nature.com/articles/nbt.3519
      - https://pachterlab.github.io/sleuth/about
  - by spike-in  Jiang,L., Schlesinger,F., Davis,C.A., Zhang,Y., Li,R., Salit,M., Gingeras,T.R. and Oliver,B. (2011) Synthetic spike-in standards for RNA-seq experiments. Genome Res., 21, 1543–1551.
  - count-based
    - analysis only at the gene level
      - cite:pachter2015
    - edgeR
  - TPM
  - FPKM
  - https://europepmc.org/article/PMC/6058388
  - abundance
    - http://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html#introduction
  - cite:pachter2015
- Bullard,J.H., Purdom,E., Hansen,K.D. and Dudoit,S. (2010) Evaluation of statistical methods for normalization and differential expression in mRNA-Seq experiments. BMC Bioinformatics, 11, doi:10.1186/1471-2105-11-94.
- cite:peixoto2015
- https://www.the-scientist.com/news-opinion/technical-bias-widespread-in-rna-seq-datasets-66766?utm_campaign=TS_DAILY%20NEWSLETTER_2019&utm_source=hs_email&utm_medium=email&utm_content=79783848&_hsenc=p2ANqtz-_TvsXNNQSHgNWR9yE8Q6vZP04TRSx3T-W0rLk0NLucSkrSI0wQ0GDQ2L1Vs_Jf1x1_21ou8yVG1ur9nXtXbBYl91OsUg&_hsmi=79783848
- normalization to control for systemic error / bias
- shrinkage: pooling information across genes to obtain reliable variance estimates for small sample sizes, cite:pachter2015
- feature counting
  - HTSeq
  - featureCounts
- https://www.slideshare.net/mobile/jakonix/part-1-of-rnaseq-for-de-defining-the-goal
- https://www.biostars.org/p/404211/#:~:text=What%20is%20the%20difference%20between,as%20fragment%20length%20distribution%20etc.
- https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html
- [[*rsem][rsem]]
- https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/
- https://github.com/griffithlab/rnaseq_tutorial/wiki/Transcript-Assembly-Merge
- https://www.reddit.com/r/bioinformatics/comments/i6qhrx/samples_from_different_studies_de_analysis_help/
- https://www.nature.com/articles/ng1032z.pdf
- evaluation
  - linearity
    - https://link.springer.com/article/10.1186/s12859-017-1526-y
  - https://bioinformatics.stackexchange.com/questions/92/confirm-success-or-failure-of-rna-seq-normalization
    - https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4404308/
- https://groups.google.com/forum/m/#!topic/rsem-users/GRyJfEOK1BQ
- https://www.biostars.org/p/68126/
- https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-3-r25
- https://cgatoxford.wordpress.com/2016/08/17/why-you-should-stop-using-featurecounts-htseq-or-cufflinks2-and-start-using-kallisto-salmon-or-sailfish/
- isoform abundance
  - https://cgatoxford.wordpress.com/2016/08/17/why-you-should-stop-using-featurecounts-htseq-or-cufflinks2-and-start-using-kallisto-salmon-or-sailfish/
- https://www.biostars.org/p/329625/
- sources of bias
  -
- https://htseq.readthedocs.io/en/release_0.11.1/tour.html#counting-reads-by-genes
- RSEM
  - https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-323
***** featureCounts in Rsubread
:PROPERTIES:
:ID:       0FB3EE5E-DD17-4438-BF35-011555A41DF9
:END:

[[http://www.rdocumentation.org/packages/Rsubread/functions/featureCounts][nice featureCounts manual]]

***** [#Y] summarizeOverlaps in GenomicAlignments
:PROPERTIES:
:ID:       16408005-913F-4675-BA61-B7594A3FB4EC
:END:

***** HTseq-count
:PROPERTIES:
:ID:       E2262D77-424F-4943-8DA4-5649B2F0B4B2
:END:

[[file:~/Box%20Sync/comp_bio/Anders,15%20HTSeq%20a%20python%20framework%20to%20work%20with%20high%20throughput%20sequencing%20data.pdf][Anders,15 HTSeq a python framework to work with high throughput
sequencing data]]

Not on cluster
- create R script to run stats on generic output file
- % no feature, ambiguous, too low qual, not aligned, alignment not
  unique

****** Prerequisites
:PROPERTIES:
:ID:       D5102869-0833-4513-A716-69D44892FDDF
:END:

- samtools sort (to postion)

****** [[http://www-huber.embl.de/HTSeq/doc/count.html#count][website]]
:PROPERTIES:
:ID:       6B7F3881-DA7D-4700-BC30-9EF83150D916
:END:

htseq-count [options] <alignment_file> <gff_file>

simple code: htseq-count -m union -r pos -i gene_name -a 10 --stranded=no agcgccg.bam Mus_musculus.GRCm38.84.gtf > output.counts

Important options
--stranded=no (defualt is yes)
--format bam (defualt is sam)
  - for paired end, order is important
    - do samtools sort (default) to put in position order
    - in HTSeq-counts, specify -r pos
    -a 10 (minimum quality, 10 is default)
    -m default is union, alternate intersection-strict and intersection-nonempty

    Why union is default: consider two genes, which share a stretch of common sequence such that for a read mapping to this stretch, the aligner cannot decide which of the two genes the read originated from and hence reports a multiple alignment. If we discard all such reads, we undercount the total output of the genes, but the ratio of expression strength (the “fold change”) between samples or experimental condition will still be correct, because we discard the same fratcion of reads in all samples. On the other hand, if we counted these reads for both genes, a subsequent diffential-expression analysis might find false positives:


  followed by the special counters, which count reads that were not counted for any feature for various reasons

  [[http://www-huber.embl.de/HTSeq/doc/counting.html][algorithm details]]


***** Expression Level Abundance, Normalization Units
:PROPERTIES:
:ID:       D4BFCCCA-8BBC-4D23-B71B-F8C2ED6287E3
:END:
https://ro-che.info/articles/2016-11-28-rna-seq-normalization
R/FPKM and C/TPM are "local" to the sample while TMM must be recalculated with different sample set

****** Raw- required input for DE analyses
:PROPERTIES:
:ID:       82A81565-9333-4E40-806E-474399BD73C4
:END:

****** R/FPKM
:PROPERTIES:
:ID:       9D6324B7-D9A2-44ED-A54F-B0C0C8DCECE2
:END:
https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/
- normalization for library size (i.e. sequencing depth) and transcript length
- use to
  - compare a gene across samples or
  - compare different genes within a sample
- see Mortazavi A., Williams B.A., McCue K., Schaeffer L., and Wold B. Mapping and quantifying mammalian transcriptomes by RNA-seq. Nat Methods 5(7):621–628, 2008.
- RPKM (reads per thousand nucleotides in transcript per million reads)
  - “For example, if a 2 kb transcript has 1000 reads and the total number of reads is 25 million, then RPKM = (1000/2)/25 = 20”
- FPKM (fragments per thousand nucleotides per million mapped reads)- equivalent of RPKM for paired-end (two reads per fragment)

****** CPM (counts per million) / TPM
:PROPERTIES:
:ID:       8C3481E5-B350-4D9D-A44C-DAB17FAEE1C6
:END:

  -   - see Wagner G.P., Kin K., and Lynch V.J. Measurement of mRNA abundance using RNA-seq data: RPKM measure is inconsistent among samples. Theory Biosci 131(4):281–285, 2012.=
  - limma voom uses a logCPM
  - corrects for transcript length distribution of different sample pools
  - TPM (transcripts per million)- divides by sum of transcript-length-normalized reads
    i.e. normalizes to transcripts as opposed to reads
    useful if average transcript length differs b/w samples

****** TMM (trimmed mean of M values)
:PROPERTIES:
:ID:       544B0EAB-4594-404B-9C83-B951E2244CFE
:END:

-attempts to normalize this problem:
[[file:work.org_imgs/20160419_132550_469_bk.png]]
- corrects for less "sequencing real estate" due to high expression per tissue
**** RNA-seq quality control
:PROPERTIES:
:ID:       e6cbde1b-22c3-4a23-b99e-d9d4fe0a6b95
:END:
- https://www.reddit.com/r/bioinformatics/comments/ufuk46/question_transcriptome_quality_assessments/

- https://bioconductor.org/packages/release/bioc/html/RUVSeq.html
- http://rseqc.sourceforge.net/
- https://www.reddit.com/r/bioinformatics/comments/flqb8e/how_do_i_differentiate_outliers_from_ingroup/
- https://www.rna-seqblog.com/an-rna-seq-qc-overview/
- https://www.biostars.org/p/368420/
- https://bioinformatics.stackexchange.com/questions/3572/strange-per-sequence-gc-content-results
- cite:wang2012


- samtools sort (to postion)

****** [[http://www-huber.embl.de/HTSeq/doc/count.html#count][website]]
:PROPERTIES:
:ID:       6B7F3881-DA7D-4700-BC30-9EF83150D916
:END:

htseq-count [options] <alignment_file> <gff_file>

simple code: htseq-count -m union -r pos -i gene_name -a 10 --stranded=no agcgccg.bam Mus_musculus.GRCm38.84.gtf > output.counts

Important options
--stranded=no (defualt is yes)
--format bam (defualt is sam)
  - for paired end, order is important
    - do samtools sort (default) to put in position order
    - in HTSeq-counts, specify -r pos
    -a 10 (minimum quality, 10 is default)
    -m default is union, alternate intersection-strict and intersection-nonempty

    Why union is default: consider two genes, which share a stretch of common sequence such that for a read mapping to this stretch, the aligner cannot decide which of the two genes the read originated from and hence reports a multiple alignment. If we discard all such reads, we undercount the total output of the genes, but the ratio of expression strength (the “fold change”) between samples or experimental condition will still be correct, because we discard the same fratcion of reads in all samples. On the other hand, if we counted these reads for both genes, a subsequent diffential-expression analysis might find false positives:


  followed by the special counters, which count reads that were not counted for any feature for various reasons

  [[http://www-huber.embl.de/HTSeq/doc/counting.html][algorithm details]]

***** Basic counting script
:PROPERTIES:
:ID:       AE616985-EA77-4CF0-8FEA-60FF56A04AC3
:END:

counting is computationally intensive- script HTseq and
summarizeOverlaps on the cluster
variabilize RNA seq workflow- make into function
**** RNA-seq exploratory data analysis
:PROPERTIES:
:ID:       77d2a559-9858-453a-83c5-a646bbf473a3
:END:
- https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html
- https://www.reddit.com/r/bioinformatics/comments/u1hsjy/great_thread_on_the_importance_of_eda_and/
- A more fundamental problem, however, arises from the high dimensional-
ity of the expression state. As the number of dimensions increases,
traditional distance metrics, such as Euclidean distance, gradually
lose their ability to distinguish the closest and most distant points18. cite:kharchenko2021


https://bookdown.org/rdpeng/exdata/k-means-clustering.html
- cite:vidman2019
Bergkvist,A., Rusnakova,V., Sindelka,R., Garda,J.M., Sjogreen,B.,
Lindh,D., Forootan,A. and Kubista,M. (2010) Gene expression
profiling–clusters of possibilities. Methods, 50, 323–335.
228. Beyene,J., Tritchler,D., Bull,S.B., Cartier,K.C., Jonasdottir,G.,
Kraja,A.T., Li,N., Nock,N.L., Parkhomenko,E., Rao,J.S. et al.
(2007) Multivariate analysis of complex gene expression and clinical
phenotypes with genetic marker data. Genet. Epidemiol., 31(Suppl.
1), S103–S109.

- cite:buschmann2016 pg 6011
- hierarchical clustering
  - eucledian
  - pearson
- partitioning
  - k-means
  - self-organizing maps
- latent variable
  - PCA
    - add scores and loadings for initial variable
- [[id:588f6536-9db9-4279-9e67-800890c2de23][LINK TO High dimension data transformation]]
***** Clustering
:PROPERTIES:
:ID:       22a6e4a4-60c2-44cf-8e5a-3ff351a02aa3
:END:
https://old.reddit.com/r/AskStatistics/comments/qcxb6c/evaluation_of_clustering_rand_index_and_nmi_i_do/
https://developers.google.com/machine-learning/clustering/clustering-algorithms
https://www.analyticsvidhya.com/blog/2016/11/an-introduction-to-clustering-and-different-methods-of-clustering/
https://www.datanovia.com/en/lessons/cluster-analysis-example-quick-start-r-code/
http://statweb.stanford.edu/~susan/courses/stats305c/examplesClusters.html
- https://mubaris.com/posts/kmeans-clustering/
https://uc-r.github.io/hc_clustering
https://www.biostars.org/p/287512/
- https://www.datanovia.com/en/blog/types-of-clustering-methods-overview-and-quick-start-r-code/

- cite:vidman2019
**** Differential gene expression
:PROPERTIES:
:ID:       ee030cef-74d4-435d-8228-b85f0e6203db
:END:
https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2021/RNAseq/Markdowns/09_Linear_Models.html
https://bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/designmatrices.html#design-matrices-with-and-without-intercept-term
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7873980/
https://www.nature.com/articles/s42003-023-04791-5

cite:buschmann2016
"y. Even though comparisons of software
performances on small RNA-Seq data are scarce, a num-
ber of independent and extensive evaluations for mRNAs
based on either synthetic data with clearly defined prop-
erties (216), or on biological datasets with validated gene
expressions (217,218) have been made recently. While it
was shown that statistical power of almost all methods is
heavily dependent on the number of samples per condition
and less on sequencing depth, the variability of expression
changes in biological datasets affects each analysis tool dif-
ferently. Outliers, ‘ON/OFF’ expression changes, where a
gene is detected in only one condition, and lopsided ex-
pression patterns, where upregulations drastically outweigh
downregulations or vice versa, influence specificity (false
positive rate) and sensitivity (false negative rate) of each
method unequally. Nonetheless, some methods appear to
capture the true expression status of small RNAs better
than others. Most independent evaluations seem to agree
that calling differential expression with SAMSeq works well
for datasets with sufficient sample sizes of 10 or more. For
smaller datasets, edgeR and especially the more conserva-
tive DESeq (or DESeq2) are found to be the methods of
choice. On"

https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7873980/
https://rnabio.org/module-03-expression/0003/06/01/DE_Pathway_Analysis/
https://www.nature.com/articles/s41598-018-29077-3
https://rnabio.org/module-03-expression/0003/05/01/Batch-Correction/
https://compgenomr.github.io/book/clustering-grouping-samples-based-on-their-similarity.html
- https://www.bioconductor.org/packages/devel/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html

- [[https://www.bioconductor.org/packages/devel/workflows/vignettes/RNAseq123/inst/doc/limmaWorkflow.html#differential-expression-analysis][easy as 1-2-3]]
https://github.com/hbctraining/DGE_workshop_salmon_online
https://f1000research.com/articles/5-1438/v2
https://www.nature.com/articles/s41598-020-74567-y#Abs1
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6954399/
- Results reporting
  - Examples
    - cite:ward2021 page 4
***** Visualizations
:PROPERTIES:
:ID:       4b2840f1-f0b3-40dc-b763-f0d2988fe146
:END:
- volcano
- heatmap
-
***** time course
:PROPERTIES:
:ID:       43f1772e-f8ce-4f6b-a4bd-d35d381b89c9
:END:
  - cite:spies2017
  - https://support.bioconductor.org/p/65676/#66860
  - https://www.biostars.org/p/138988/
  - https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#group-specific-condition-effects-individuals-nested-within-groups
  - https://hbctraining.github.io/DGE_workshop/lessons/08_DGE_LRT.html
  - http://seqanswers.com/forums/showthread.php?t=64039
  - https://www.sciencedirect.com/science/article/pii/S2001037015000392
***** experimental design
:PROPERTIES:
:ID:       4b9eee0d-4dd2-4b31-a5c2-873663443671
:END:
  - cite:mccarthy2012
  - https://angus.readthedocs.io/en/2016/_static/DifferentialExpressionBasics_NGS2016_ID.pdf
***** specialized DE datasets
:PROPERTIES:
:ID:       722cd834-2504-4ae0-84ef-58554fa4da32
:END:
- long non-coding- https://europepmc.org/article/PMC/6058388
- transcript / isoform DE
  - https://davetang.org/muse/2017/10/25/getting-started-hisat-stringtie-ballgown/
  - sleuth https://www.nature.com/articles/nmeth.4324 cite:pachter2015
  - CuffDiff- isoform count-based
***** tutorials
:PROPERTIES:
:ID:       9bd06c29-45ba-42b8-b048-27b83a0fa47d
:END:
- http://master.bioconductor.org/packages/release/workflows/html/rnaseqGene.html
- https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#multi-factor-designs
- https://bioconductor.org/packages/devel/workflows/vignettes/RNAseq123/inst/doc/limmaWorkflow.html#useful-graphical-representations-of-differential-expression-results
- [[https://github.com/griffithlab/rnaseq_tutorial/wiki/Citation][Griffith RNA-seq tutorial]]
***** statistical models
:PROPERTIES:
:ID:       8a788e6c-3899-4e6f-bf09-9c96d110aeaa
:END:
https://bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/designmatrices.html#design-matrices-with-and-without-intercept-term
- overdispersion makes working under Poisson assumption prone to high false positives b/c sampling error underestimated


Di,Y., Schafer,D.W., Cumbie,J.S. and Chang,J.H. (2011) The NBP
negative binomial model for assessing differential gene expression
from RNA-Seq. Stat. Appl. Genet. Mol. Biol., 10,
doi:10.2202/1544-6115.1637.



  - cite:huang2015
  - https://angus.readthedocs.io/en/2016/_static/DifferentialExpressionBasics_NGS2016_ID.pdf
  - https://bioinformatics-core-shared-training.github.io/RNAseq-R/slides/LinearModels.pdf
  - dispersion
    - https://support.bioconductor.org/p/75260/
  - no t-test
    - cite:jeanmougin2010
    - https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0012336
  - RNA expression varies about 10^5
  - RNA counts follow “overdispersed” poisson ( (large number of events, each event is rare) / negative binomial distribution (additional factor is dispersion)
    expression levels of same gene across tissues follow log-normal distribution
    - base distribution is Poisson (b/c large number of events, each event is rare), variance = mean
    - shot noise- noise from sampling
    - zero inflation- many zero counts
    - negative binomial (overdispersed Poisson), variance = mean + (1/r)*u^2, used by
      - DESeq2 (can also do continuous values like limma)
      - EdgeR
***** Methods and tools
:PROPERTIES:
:ID:       68BB9D3D-5879-4F2E-904B-54F70EAD2B5A
:END:
- count-based
  - negative binomial
    - [[id:32BB77F2-0283-4A52-86AC-CDF2CB6A7315][edgeR]], dispersion by weighted likelihood
    - [[id:2F642F69-A9C8-46DA-BE31-783AC8D4F4B6][DESeq2]], dispersion by modeling all genes via regression
- [[id:F9684168-A438-4F5E-81D6-D645B52B4074][limma voom]], voom transforms read count
- Bayesian
  - BaySeq
  - BitSeq
  - ebSeq
- Non-parametrics- require 5-10 replicates per group
  - SAMSeq
  - NOISeq
- methods comparisons
  - cite:fonseca2014
  - cite:huang2015
  - cite:seyednasrollah2015
  - cite:tang2015
  - cite:rapaport2013
  - https://www.nature.com/articles/s41598-020-59516-z#data-availability
  - comparison
    - https://academic.oup.com/bib/article/16/1/59/240754#26957696
- de novo
  - https://informatics.fas.harvard.edu/best-practices-for-de-novo-transcriptome-assembly-with-trinity.html
***** outputs
:PROPERTIES:
:ID:       48db2c9c-5d17-4438-aa40-43816245c490
:END:
  - log fold change
    - https://www.biostars.org/p/101727/


***** Notes
:PROPERTIES:
:ID:       558011C7-FB50-435E-B130-E3485CC25B90
:END:
- cite:anders2010
- https://pubmed.ncbi.nlm.nih.gov/22931062/
- bench methods
  - cite:parekh2016


**** [[id:91d46e60-eec3-48c7-97f1-b53eb1559b3f][Co-expression]]
**** RNA-seq visualization
:PROPERTIES:
:ID:       b1351d51-3ac1-4193-b944-850c44a903fd
:END:
- http://gepia.cancer-pku.cn/
- [[id:DB8463CB-97CB-49E6-8EDD-A5ECA2F140BF][visualization tools]]
- examples
  - [[https://www.cell.com/cell-reports/pdf/S2211-1247(20)30669-0.pdf][good seq analysis and visual example]]
- volcano
  - https://www.biostars.org/p/203876/
- heatmaps
  - http://compbio.ucsd.edu/making-heat-maps-r/
  - https://www.rdocumentation.org/packages/gplots/versions/3.0.3/topics/heatmap.2
- mapped reads were visualized by SeqMonk[[file:org.org_imgs/20161220_180037_4849srR.png]]
**** Methods by RNA population
:PROPERTIES:
:ID:       dfc3be20-ffb1-4587-86e8-362e2cecf8bb
:END:
***** exRNA
:PROPERTIES:
:ID:       d1b1f9a2-b228-461f-8553-15adf8426d28
:END:
- [[cite:&hulstaert2021]]
- [[cite:&hulstaert2020]]
- [[cite:&exrnaqc2021]]
- spike-in use
  - cite:godoy2019
  - cite:giraldez2018

- cite:godoy2018
- cite:alexander2022
- spike ins use
  - cite:godoy2019
  - cite:giraldez2018
- [[id:fd462ab0-6aaf-4732-953b-141ad5e2ee72][small RNA sequencing]]
- https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9358052/
- https://www.biorxiv.org/content/10.1101/113050v1.full
- cite:vorperian2022
- https://www.sciencedirect.com/science/article/pii/S0092867419301679?via%3Dihub#sec4
- https://www.sciencedirect.com/science/article/pii/S2211124720315412?via%3Dihub
- cite:yeri2018


****** [[id:80cff175-c41d-42d4-aaa0-39546e0a6f1d][Plasma exRNA (cfRNA) extraction and isolation]]
:PROPERTIES:
:ID:       46833b57-2d7f-4f7d-9cef-98edde71a09f
:END:
****** [[id:51e5d70c-4a62-4960-880e-56f050194a0a][exRNA Biology]]
:PROPERTIES:
:ID:       694ecf28-e611-4de0-948b-5df35b876efe
:END:
****** [[id:b2873d1a-837b-4927-b8cf-8a27e42d7bea][small RNA extraction]]
:PROPERTIES:
:ID:       9ab4e370-c395-42ca-9a38-191fee9c89bb
:END:
****** [[id:71289eeb-a929-44bd-a46e-a0c3dcd7a8f4][small RNA library preparation]]
:PROPERTIES:
:ID:       ed6f2c9d-1ef5-4274-a8d5-fb3e55464c4d
:END:
****** [[id:fd462ab0-6aaf-4732-953b-141ad5e2ee72][small RNA sequencing]]
:PROPERTIES:
:ID:       7da11c3b-384c-4f1f-87f5-e13975b8b925
:END:
****** QC
:PROPERTIES:
:ID:       59ec7abc-dd76-4c09-b84a-75e1e08cfcd3
:END:
- cite:lee2016exrna
****** Complete pipelines
:PROPERTIES:
:ID:       71bf0765-1582-4d5e-b11e-64317d09dfed
:END:
- https://exrna.org/resources/software/
- pipelines- [[id:1db0ffb6-25b7-4c97-9aa0-263b8c96373b][exceRpt]]
- https://github.com/oncornalab/rnaexome from [[cite:&hulstaert2021]]
****** By transcript and sequencing
:PROPERTIES:
:ID:       f8d7339e-5a7f-4b95-bcf3-b1bf266ae588
:END:
- single cell- https://www.nature.com/articles/s41587-021-01188-9
****** By fluid
:PROPERTIES:
:ID:       6e1ef9db-2ac5-4156-9eab-5e99a2875cdc
:END:
- cite:godoy2018
******* Plasma
:PROPERTIES:
:ID:       f4018b42-1681-415d-a96f-1ce8542b1b97
:END:
- non-human plasma small RNA
  - cite:vliet2017
- human plasma small RNA
  - cite:danielson2018
  - cite:godoy2018
  - cite:max2018
  - cite:yeri2018
  - cite:danielson2017
  - cite:juzenas2017
  - cite:yeri2017
  - cite:freedman2016
- plasma transcriptome RNA
  - cite:zhao2018
  - cite:chen2019exrna
******* urine
:PROPERTIES:
:ID:       a454bd50-ba17-4fb1-bfc8-350826287b67
:END:
  - https://pubmed.ncbi.nlm.nih.gov/28303895/
******* salivary
:PROPERTIES:
:ID:       37b4a512-c6cc-46d0-8a78-3bb4e2315acb
:END:
  - https://pubmed.ncbi.nlm.nih.gov/28303895/
  - - https://www.ncbi.nlm.nih.gov/pubmed/28961734
****** General reference
:PROPERTIES:
:ID:       536e1127-fa49-4bc5-bb16-47cfcc81b1ad
:END:
- Extracellular RNA Communication Consortium
  - https://www.sciencedirect.com/science/article/pii/S0092867419302855?via%3Dihub
- - cell of origin -
  - https://www.biorxiv.org/content/10.1101/2021.05.05.441859v1.full
  - https://www.nature.com/articles/s41587-021-01188-9
- cite:alexander2022
- pipelines- [[id:1db0ffb6-25b7-4c97-9aa0-263b8c96373b][exceRpt]]

***** small RNA sequencing
:PROPERTIES:
:ID:       fd462ab0-6aaf-4732-953b-141ad5e2ee72
:END:
- cite:giraldez2018
- https://link.springer.com/protocol/10.1007%2F978-1-4939-9042-9_4
- cite:buschmann2016
****** General reference
:PROPERTIES:
:ID:       65a85f59-afd5-4c6b-a785-9b6269cc7a51
:END:
- miRQC cite:mestdagh2014
- cite:mehta2014
- [[https://www.biostars.org/t/smallrna/?sort=votes&limit=all%20time][biostars small RNA]]
- cite:korpelainen2015 chapter 12
- best practices
  - https://www.encodeproject.org/data-standards/rna-seq/small-rnas/


****** QC
:PROPERTIES:
:ID:       fc7aeeb1-d387-4379-be37-46ebf1244cd7
:END:
https://exrna.org/resources/data/data-quality-control-standards/
****** Pre-processing
:PROPERTIES:
:ID:       0c7d62b9-6ef4-43d1-8e21-c3aa536125bd
:END:
- read filtering
  - adapter trimming
  - quality filtering
  - read length trimming
    - remove at less than 16 nt (see cite:buschmann2016 ref 157,158)
    - these are adapter dimers and degraded RNA
- QC
  - filtered reads should have miRNA 21-23 nt peak and piRNA 30-32 nt peak
  - overrepresented sequence- k-mer distribution- see cite:buschmann2016 pg 6007
- [[id:1EE731A3-83DD-46AF-B988-2311307C335F][LINK TO Raw sequence QC and processing]]
****** [[id:51e5d70c-4a62-4960-880e-56f050194a0a][LINK TO Extracellular RNA biology]]
:PROPERTIES:
:ID:       49a0e52b-0bda-4237-8bab-ff2db54cf26d
:END:
****** Design
:PROPERTIES:
:ID:       4e8a8446-e84d-4c4f-8bd8-45ee611337b5
:END:
- read length
- spike-ins
  - https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4538800/
- validation technologies
  - cannot distinguish isomiRs
  - qPCR expected to correlate >.9 w/ seq expression > 2^8 cite:yeri2018
  - qPCR more concordant than EdgeSeq or FirePlex cite:yeri2018
- what depth? https://www.researchgate.net/post/how_many_million_reads_for_miRNA_discovery
- reads per sample
  - https://www.qiagen.com/us/resources/faq?id=a119d5f0-0a2a-4612-828e-d8fa3cc37891&lang=en
****** [[id:b2873d1a-837b-4927-b8cf-8a27e42d7bea][LINK TO small RNA extraction]]
:PROPERTIES:
:ID:       dfc3a8a2-e4c4-4043-af71-2db93128c18e
:END:
****** [[id:4E23127E-4ED6-4802-A03C-4103AA8CAF29][LINK TO RNA Quality and Quantification]]
:PROPERTIES:
:ID:       4df90313-39e4-44ed-83cd-b98453e18358
:END:
****** small RNA library preparation
:PROPERTIES:
:ID:       71289eeb-a929-44bd-a46e-a0c3dcd7a8f4
:END:
- cite:wong2019
- cite:coenen-stass2018
- cite:qiagen2019biofluid_mirna
- Truseq small finds more tRNA fragments than other kits cite:yeri2018
- Bioscientific NextFlex maps more reads to miRNA than other kits cite:yeri2018 (~75% vs 65% for truseq in plasma)
- read diversity correlates to input volume, but b/w 550 and 650 at >1 RPM at volumes 200 uL to 5 mL cite:yeri2018 8c
- https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6161688/
- cite:buschmann2016
- cite:illumina2016truseq_small
- 10-50 ng isolated or 1 ug total https://support.illumina.com/content/dam/illumina-support/documents/documentation/chemistry_documentation/samplepreps_truseq/truseqsmallrna/truseq-small-rna-library-prep-kit-reference-guide-15004197-02.pdf
- library prep - cite:wright2019
- https://dx.doi.org/10.1186%2Fs12864-018-4726-6
- https://dx.doi.org/10.1186%2Fs12864-018-4491-6
- small RNA library preparation cite:illumina2016
- kit consistency required- see cite:yeri2018 8b
- with low input, use same kit for reproducibility cite:yeri2018
- cite:dard-dascot2018
****** [[id:3BC41E6A-00D3-4498-A48C-463F68CFD76B][LINK TO Massively parallel sequencing]]
:PROPERTIES:
:ID:       6584c294-78b9-42e5-a25b-28a069e112d8
:END:
****** Bioinformatics
:PROPERTIES:
:ID:       f46e278c-dd38-4603-9fa2-a973823e1333
:END:
- alignment to miRbase needs to be more conservative to avoid mismatches
- cite:buschmann2016
- https://link.springer.com/protocol/10.1007%2F978-1-4939-9042-9_4
******* Pre-processing
:PROPERTIES:
:ID:       0c7d62b9-6ef4-43d1-8e21-c3aa536125bd
:END:
- read filtering
  - adapter trimming
  - quality filtering
  - read length trimming
    - remove at less than 16 nt (see cite:buschmann2016 ref 157,158)
    - these are adapter dimers and degraded RNA
- QC
  - filtered reads should have miRNA 21-23 nt peak and piRNA 30-32 nt peak
  - overrepresented sequence- k-mer distribution- see cite:buschmann2016 pg 6007
- [[id:1EE731A3-83DD-46AF-B988-2311307C335F][LINK TO Raw sequence QC and processing]]
******* Alignment
:PROPERTIES:
:ID:       d657a691-e338-44ad-96e4-47b4b18a03d2
:END:
- alignment to miRbase needs to be more conservative to avoid mismatches
******* QC
:PROPERTIES:
:ID:       fc7aeeb1-d387-4379-be37-46ebf1244cd7
:END:
https://exrna.org/resources/data/data-quality-control-standards/
******* [[id:853dd34c-27be-4be8-9794-19f04a1e22b0][LINK TO Quantification and normalization]]
:PROPERTIES:
:ID:       9a349ccd-3d29-46b7-b4fa-71693d32c674
:END:
******* [[id:ee030cef-74d4-435d-8228-b85f0e6203db][LINK TO Differential expression]]
:PROPERTIES:
:ID:       56d53b46-2a62-4777-b5c7-eedbc306d4b4
:END:
cite:buschmann2016
"y. Even though comparisons of software
performances on small RNA-Seq data are scarce, a num-
ber of independent and extensive evaluations for mRNAs
based on either synthetic data with clearly defined prop-
erties (216), or on biological datasets with validated gene
expressions (217,218) have been made recently. While it
was shown that statistical power of almost all methods is
heavily dependent on the number of samples per condition
and less on sequencing depth, the variability of expression
changes in biological datasets affects each analysis tool dif-
ferently. Outliers, ‘ON/OFF’ expression changes, where a
gene is detected in only one condition, and lopsided ex-
pression patterns, where upregulations drastically outweigh
downregulations or vice versa, influence specificity (false
positive rate) and sensitivity (false negative rate) of each
method unequally. Nonetheless, some methods appear to
capture the true expression status of small RNAs better
than others. Most independent evaluations seem to agree
that calling differential expression with SAMSeq works well
for datasets with sufficient sample sizes of 10 or more. For
smaller datasets, edgeR and especially the more conserva-
tive DESeq (or DESeq2) are found to be the methods of
choice. On"
******* Functional analysis
:PROPERTIES:
:ID:       e597d527-5538-4f69-bd28-a24a70969317
:END:
cite:buschmann2016 pg6011
******* Integrated tools
:PROPERTIES:
:ID:       07055c47-6e24-4f59-bce4-0ab91a7f5377
:END:
- miRDeep2 / miRDeep
  - cite:friedlander2012
- exceRpt / genboree
  - http://github.gersteinlab.org/exceRpt/
  - using
    - cite:godoy2018
    - cite:freedman2016
  - http://genboree.org/java-bin/workbench.jsp
- miRge / miRge2 WOS 29 times cited
  - https://mirge3.readthedocs.io/en/latest/quick_start.html
  - https://github.com/mhalushka/miRge3.0
  - cite:lu2018mirge
  - https://github.com/mhalushka/miRge
    - https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-018-2287-y
- sRNAbench
  - https://academic.oup.com/nar/article/47/W1/W530/5494756
  - https://bioinfo2.ugr.es/ceUGR/srnabench/
  - sRNAtoolbox
    - https://arn.ugr.es/srnatoolbox/srnabench/
    - https://arn.ugr.es/srnatoolbox/static/sRNAtoolbox_manual.pdf
****** Applications
:PROPERTIES:
:ID:       55212621-170f-41aa-b9b5-b1c00385ffe6
:END:
******* Extracellular RNA bioinformatics
:PROPERTIES:
:ID:       cbf67ef7-39ec-45ec-b599-cdc0e8590cf7
:END:
- spike ins use
  - cite:godoy2019
  - cite:giraldez2018
- pipelines- [[id:1db0ffb6-25b7-4c97-9aa0-263b8c96373b][exceRpt]]
******** [[id:71289eeb-a929-44bd-a46e-a0c3dcd7a8f4][small RNA library preparation]]
:PROPERTIES:
:ID:       2e392314-afcf-49b1-8dea-78fb0bb8f18e
:END:
******** [[id:b2873d1a-837b-4927-b8cf-8a27e42d7bea][small RNA extraction]]
:PROPERTIES:
:ID:       1ae87816-2819-40d0-8c8e-b039d3885d74
:END:
******** By transcript and sequencing
:PROPERTIES:
:ID:       f08d4d6f-a4a1-4c0a-b8f9-d722d90a3aa7
:END:
- single cell- https://www.nature.com/articles/s41587-021-01188-9
******** By fluid
:PROPERTIES:
:ID:       a5bfc41c-ebcb-4f31-bf8c-3099e5ad170b
:END:
- cite:godoy2018
********* Plasma
:PROPERTIES:
:ID:       c3ec22fa-fd6b-4c74-b4c6-71aa346819c2
:END:
- non-human plasma small RNA
  - cite:vliet2017
- human plasma small RNA
  - cite:danielson2018
  - cite:godoy2018
  - cite:max2018
  - cite:yeri2018
  - cite:danielson2017
  - cite:juzenas2017
  - cite:yeri2017
  - cite:freedman2016
- plasma transcriptome RNA
  - cite:zhao2018
  - cite:chen2019exrna
********* urine
:PROPERTIES:
:ID:       554d0ae8-d299-4d9e-83fa-26398f75db43
:END:
  - https://pubmed.ncbi.nlm.nih.gov/28303895/
********* salivary
:PROPERTIES:
:ID:       ef77efdc-a0e8-4ad3-b26f-9c9faf35241d
:END:
  - https://pubmed.ncbi.nlm.nih.gov/28303895/
  - - https://www.ncbi.nlm.nih.gov/pubmed/28961734
******* miRNA-seq
:PROPERTIES:
:ID:       903f32d8-bd13-496c-8d5c-26a63455936a
:END:
:LOGBOOK:
CLOCK: [2021-01-27 Wed 13:53]--[2021-01-27 Wed 13:57] =>  0:04
:END:
- spike-ins
  - https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4538800/
- reads per sample
  - https://www.qiagen.com/us/resources/faq?id=a119d5f0-0a2a-4612-828e-d8fa3cc37891&lang=en
- cite:tam2015
- https://www.nature.com/articles/srep05150#Sec11
- http://www.mircode.org/faq.php
- https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4652620/
- https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6161688/
- [[id:5d9593fd-f083-450f-a510-449aa15ae670][LINK TO miRNA biology]]
- [[id:95b3c7f3-3b84-45b1-b8bf-5e2e6e2a6f88][LINK TO miRNA data]]
- spike-ins
  - cite:qiagen2018mirnaspike cite:exiqon2016exiseq
  - cite:locati2015
  - https://genome.cshlp.org/content/21/9/1543.long
- processes
  - preprocessing
    - read collapse just speads up downstream
      - see for read collapse https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0143066
  - QC
    - library complexity
    - https://arn.ugr.es/mirnaqc/
  - normalization
    - https://academic.oup.com/bib/article/16/6/950/226343
    - https://rnajournal.cshlp.org/content/18/6/1279.long
  - functional analysis
    - https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4489228/
  - target identification
    - http://ophid.utoronto.ca/mirDIP/
  - [[id:82aafb6f-63a8-4ab7-867a-9add0e0c224a][LINK TO Integrated miRNA-mRNA analysis]]
- tools
  - manatee cite:handzlik2020
  - tool comparisons
    - https://www.frontiersin.org/articles/10.3389/fgene.2018.00022/full
    - cite:chen2019mirna
- general reference
  - cite:pritchard2012
    - citing, times cited WOS:
      - cite:braunwald2013
      - cite:dong2013
      - cite:schwarzenbach2014
      - " Whereas high-speed centrifugation (14,000–16,000 g) can remove living cells and cell debris, separating exosomes and small shedding vesicles requires ultracentrifugation (>110,000 g).30 Blood cell count has a significant effect on miRNA expression levels in plasma;31 therefore, samples from patients with an inflammatory process, who should have high white blood cell count, should be excluded from analysis. Haemolysis is also known to markedly influence miRNA expression patterns, because of a contamination of abundant intracellular miRNAs.31"
  - [[http://apps.pathology.jhu.edu/mirnaprogram/links/http://apps.pathology.jhu.edu/mirnaprogram/links/][Johns Hopkins microRNA Bio lab]]
  - https://www.biostars.org/t/mirna/
- protocols
  - cite:fu2018mirna
  - https://www.sciencedirect.com/science/article/pii/S266591312030131X#sec3
****** General reference
:PROPERTIES:
:ID:       65a85f59-afd5-4c6b-a785-9b6269cc7a51
:END:
- miRQC cite:mestdagh2014
- cite:mehta2014
- [[https://www.biostars.org/t/smallrna/?sort=votes&limit=all%20time][biostars small RNA]]
- cite:korpelainen2015 chapter 12
- best practices
  - https://www.encodeproject.org/data-standards/rna-seq/small-rnas/


****** Integrated miRNA-mRNA analysis
:PROPERTIES:
:ID:       82aafb6f-63a8-4ab7-867a-9add0e0c224a
:END:
- [[id:903f32d8-bd13-496c-8d5c-26a63455936a][LINK TO miRNA-seq]]
- example papers
  - https://bmccancer.biomedcentral.com/articles/10.1186/s12885-019-5983-8#Sec2
  - https://www.oncotarget.com/article/8713/text/
  - cite:li2017mirna
**** Methods by tissue source or isolation
:PROPERTIES:
:ID:       4d4a89d1-e039-4a89-b968-81c80f38d884
:END:
***** Salivary
:PROPERTIES:
:ID:       4be084dc-50ab-4a56-8ab7-fa180b7c5d83
:END:
  - https://pubmed.ncbi.nlm.nih.gov/28303895/
  - - https://www.ncbi.nlm.nih.gov/pubmed/28961734

***** Urine
:PROPERTIES:
:ID:       a6e713dd-b989-4991-8a2b-938995198599
:END:
  - https://pubmed.ncbi.nlm.nih.gov/28303895/

***** Plasma
:PROPERTIES:
:ID:       8eaeeb06-f610-4d9d-80e1-06c148a85646
:END:
- non-human plasma small RNA
  - cite:vliet2017
- human plasma small RNA
  - cite:danielson2018
  - cite:godoy2018
  - cite:max2018
  - cite:yeri2018
  - cite:danielson2017
  - cite:juzenas2017
  - cite:yeri2017
  - cite:freedman2016
- plasma transcriptome RNA
  - cite:zhao2018
  - cite:chen2019exrna

***** ffpe
:PROPERTIES:
:ID:       632e82ae-a39d-4953-a642-85445ca8a7f6
:END:
  - cite:kashofer2013
  - FFPE Pre-analytic variability
    - cite:ahlfen2007rnaqual
      - storage conditions
      - cold ischemia time
      - fixation
        - type
        - time
      - sample thickness
    - biospecimens.cancer.gov/researchnetwork
    - cite:unlockingQiagen
    - https://www.ncbi.nlm.nih.gov/pubmed/25357115
    - By effect
      - Yield
        - - prolonged fix reduces yield- Bresters D, Schipper ME, Reesink HW, Boeser-Nunnink BD, Cuypers HT. The duration of fixation influences the yield of HCV cDNA-PCR products from formalin-fixed, paraffin- embedded liver tissue. J Virol Methods 1994;48:267–72.
        - cite:webster2015mining longer fixation times decrease RNA yields from paraffin-embedded tissues (herein 3 weeks in formalin resulted in half the RNA yield/sec- tion compared with 18 h) and increase RNA input needed during library preparation (see methods). This latter effect is likely due to ‘non-functional’ RNA (eg, adducts or damaged bases) that could be potentially mitigated by treatments like PreCR or other enzymatic and chemical methods.
      - At microarray expression
        - cite:kashofer2013quality
    - "Formalin fixation introduced major changes into microarray gene expression data and led to marked gene-to-gene variations in delta-ct values of qRT-PCR. We found that qRT-PCR efficiency and gene-to- gene variations were mainly attributed to differences in the efficiency of cDNA synthesis as the most sensitive step. These differences could not be reliably detected by quality assessment of total RNA isolated from formalin-fixed tissues by electrophoresis or spectrophotometry. "
    - At sequencing
  - cite:graw2015
  - Past expression tech
    - goals were biomarker discovery and patient stratification
    - cite:Zhao_2014 manifests its superiority in terms of the dynamic range, sampling depth, and has independence from pre-existing sequence information [1,2].
    - cite:webster2015mining When compared with conventional DNA microarrays, RNA-seq enriches for many additional fragments as it is not restricted to predefined probes and has (in principle) no limitations to dynamic range (Li et al., 2014b; SEQC/MAQC-III Consortium, 2014).
    - - https://www.ncbi.nlm.nih.gov/pubmed/14695316
    - - [[https://www.ncbi.nlm.nih.gov/pubmed/15509548][DASL]]
    - https://www.ncbi.nlm.nih.gov/pubmed/21347257
  - FFPE microarray and RT-PCR was useful
    - AlloMap is a non-invasive gene expression-based blood test that is used to manage the clinical care of heart transplant recipients, providing a quantified score for the risk of rejection based on the measurement of expression of 20 genes, a subset of which are related to immune system activation and signalling15,16.
    - - cite:penland2007rna
    - - - https://www.ncbi.nlm.nih.gov/pubmed/11568892/
    - Mittempergher, L., de Ronde, J.J., Nieuwland, M., Kerkhoven, R.M., Simon, I., Th Rutgers, E.J., Wessels, L.F., Van't Veer, L.J., 2011. Gene expression profiles from formalin fixed paraffin embedded breast cancer tissue are largely comparable to fresh fro- zen matched tissue. PLoS One 6, e17163.
  - But this expression tech is inferior
    - In other work14, systematic evalu- ation of RNA-seq-based and microarray-based classifiers found that RNA-seq outperformed arrays in characteriz- ing the transcriptome of cancer and performed similarly to arrays in clinical endpoint prediction.
    - does not depend on genome annotation, or on predesigned species-specific or tran- script-specific probes, for transcript measurement.
    - microarray limited application to clinic- lthough powerful, microarray-based assays can have limitations in some environments, such as those related to laboratory-to-laboratory variation in sample preparation that can affect reproducibility. Moreover, for some applications, microarray signal-to-noise ratios can affect the limit of detection. Interestingly, a number of additional cancer multi-gene profile tests are clinically available, such as OncoTypeDX (Genome Health)135 for breast cancer recurrence risk and Prolaris (Myriad)136 for prostate cancer aggressiveness. These tests are based on qRT-PCR technologies, rather than microarrays, largely owing to the belief that qRT-PCR is more reliable, reproducible, sensitive and accurate.
    - " qRT-PCR is generally deemed the ‘gold standard’ method for measuring transcript levels, particularly in the clinical space; however, there are a number of limitations, including the fact that although it is a high sample throughput technology, relatively few markers or measurements can be made in a single assay. "
    - rnaseq superior for
      - snvs
      - indels
      - isoforms
      - splice vars
      - chimerics
        - fusions
    - semiquan
    - no SNV discovery
    - no fusion
    - "RNA-seq provides much more detailed information on specific transcript expression patterns in clinico-genetic neuroblastoma subgroups than microarrays. " cite:zhang2015comparison
**** General
:PROPERTIES:
:ID:       34f17ca7-009c-4110-b6a9-6bfbf7c8a612
:END:
- my RNA-seq work
  - [[file:~/org/career.org::*high-grade head and neck characterization][high-grade head and neck characterization]]
  - [[id:e7a8362d-f4cf-45ec-a3f8-91b0ffbdc500][ffpe rna-seq project]]
  - [[file:~/data/career/pathology/molecular-pathology/ffpe-rna/ffpe-rnaseq.org]]
- [[https://www.nature.com/subjects/rna-sequencing/nmeth][Nature Methods RNA sequencing]]
- tool comparisons cite:sahraeian2017
- data
  - public RNA sequencing data
    - [[https://webshare.bioinf.unc.edu/public/mRNAseq_TCGA/UNC_mRNAseq_summary.pdf][TCGA RSEM]]
      - https://support.bioconductor.org/p/131066/
      - http://zyxue.github.io/2017/06/02/understanding-TCGA-mRNA-Level3-analysis-results-files-from-firebrose.html
      - https://bsbludwig.com/post/94066296740/what-do-tcgas-rnaseq-files-actually-show
      - http://seqanswers.com/forums/showthread.php?t=42911
      - https://support.bioconductor.org/p/65890/
  - see [[http://www.cbil.upenn.edu/BEERS/][BEERS]] for generating fake RNA seq data

- cite:korpelainen2015
  - [[http://rnaseq-book.blogspot.fi/][Korpelainen,15 web companion]]
- [[http://rnaseq.uoregon.edu/][RNAcyclopedia]]
- comparisons cite:poplawski2015
- [[https://hub.docker.com/r/mgibio/rnaseq/dockerfile][mgi rna-seq dockerfile]]
**** Nacent RNA sequencing
:PROPERTIES:
:ID:       7b2235be-1f01-4e51-9de9-adbdfc016b36
:END:
:LOGBOOK:
CLOCK: [2022-06-06 Mon 10:45]--[2022-06-06 Mon 11:25] =>  0:40
CLOCK: [2022-06-03 Fri 08:56]--[2022-06-03 Fri 09:31] =>  0:35
:END:
- https://www.scopus.com/results/results.uri?sort=plf-f&src=s&st1=%22precision+run-on+sequencing%22+OR+%22PRO-seq%22&sid=eeff59adf884bc66d7f41fb65fee374c&sot=b&sdt=b&sl=57&s=TITLE-ABS-KEY%28%22precision+run-on+sequencing%22+OR+%22PRO-seq%22%29&origin=searchbasic&editSaveSearch=&yearFrom=Before+1960&yearTo=Present
  - cite:mahat2016
    - https://www.scopus.com/results/citedbyresults.uri?sort=cp-f&cite=2-s2.0-84979882531&src=s&nlo=&nlr=&nls=&imp=t&sid=83fb718ed42783d656150cb229690a68&sot=cite&sdt=a&sl=0&origin=resultslist&offset=1&txGid=80fce121791a099b206e9017b82f705f&isValidNewDocSearchRedirection=false
      - https://www.nature.com/articles/nature20149
      - https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1870-5
      - https://www.nature.com/articles/s41467-017-00151-0
  - https://www.nature.com/articles/nprot.2016.086
- https://genome.cshlp.org/content/27/11/1816
- https://www.scopus.com/results/results.uri?cc=10&sort=cp-f&src=s&st1=%22nascent+RNA+sequencing%22&nlo=&nlr=&nls=&sid=a567d38581c262627b32ded722d56613&sot=b&sdt=b&sl=39&s=TITLE-ABS-KEY%28%22nascent+RNA+sequencing%22%29&ss=cp-f&ps=r-f&editSaveSearch=&origin=resultslist&zone=resultslist
  - cite:core2008
  - cite:henriques2018
**** Integrating results across RNA-seq datasets
:PROPERTIES:
:ID:       6f5a8723-50ba-463d-86c4-a322aa7058bb
:END:
- cite:rajkumar2015
- https://www.nature.com/articles/s42003-023-04791-5
**** Writeup
:PROPERTIES:
:ID:       02865379-2368-4e54-aea9-3162be7e42cf
:END:
- https://www.nature.com/articles/s42003-023-04791-5
**** General reference
:PROPERTIES:
:ID:       6a25de63-b4aa-4055-b7a6-ee86157d16a6
:END:
- [cite:@li2014]
- [cite:@su2014]
- For Smyth edgeR pipeline: [cite:@chen2016dge]
**** Complete pipelines
:PROPERTIES:
:ID:       3b95fbca-3d9a-479c-9285-f8f328c5114f
:END:
- [[https://github.com/zhxiaokang/RASflow/tree/master][RASflow]]
- [[https://bitbucket.org/cfce/viper/src/master/viper.snakefile][viper]]
- [[http://bioinf.wehi.edu.au/marray/jsm2005/lab4/lab4.html][A limma workflow to try]]
- [[http://combine-australia.github.io/RNAseq-R/06-rnaseq-day1.html][another workflow example]]
- https://github.com/BIMSBbioinfo/pigx_rnaseq
***** Pipeline Comparisons
:PROPERTIES:
:ID:       7B9AD35F-1511-4CB2-BB5A-0989EFEAADF3
:END:
[[file:~/Box%20Sync/library/Fonseca,14%20RNA%20seq%20gene%20profiling%20a%20systematic%20empirical%20comparison.PDF][Fonseca,14 RNA seq gene profiling a systematic empirical comparison]]
[[file:~/Box%20Sync/library/Sigurgeirsson,14%20Analysis%20of%20stranded%20information%20using%20an%20automated%20procedure%20for%20strand%20specific%20RNA%20sequencing.pdf][Sigurgeirsson,14 Analysis of stranded information using an automated procedure for strand specific RNA sequencing]]
**** Methods for non-expression analysis
:PROPERTIES:
:ID:       8e63e018-e232-49b3-b574-892ea0bb28f0
:END:
***** cell enumeration through bulk RNA-seq
:PROPERTIES:
:CREATED:  [2020-08-05 Wed 08:36]
:ID:       1a1d5b73-8070-402a-9f26-e147d66ca081
:END:
****** CIBERSORTx
:PROPERTIES:
:ID:       bb55c105-5977-4a01-802e-4bc1838f58bb
:END:

https://www.biostars.org/p/428905/
    - raw counts will be converted to CPM in cibersortx- see https://cibersortx.stanford.edu/tutorial.php
- cibersortx
  - cite:newman2015
  - [[mu4e:msgid:BYAPR02MB5592ACF7A79C1137A6F0A974B77B0@BYAPR02MB5592.namprd02.prod.outlook.com][Re: cibersortx from docker]]
  - docker pull lyronctk/cibersortxfractions
- have access token for docker: https://cibersortx.stanford.edu/getoken.php
***** cell enumeration through bulk RNA-seq
:PROPERTIES:
:CREATED:  [2020-08-05 Wed 08:36]
:ID:       1a1d5b73-8070-402a-9f26-e147d66ca081
:END:
****** CIBERSORTx
:PROPERTIES:
:ID:       bb55c105-5977-4a01-802e-4bc1838f58bb
:END:

https://www.biostars.org/p/428905/
    - raw counts will be converted to CPM in cibersortx- see https://cibersortx.stanford.edu/tutorial.php
- cibersortx
  - cite:newman2015
  - [[mu4e:msgid:BYAPR02MB5592ACF7A79C1137A6F0A974B77B0@BYAPR02MB5592.namprd02.prod.outlook.com][Re: cibersortx from docker]]
  - docker pull lyronctk/cibersortxfractions
- have access token for docker: https://cibersortx.stanford.edu/getoken.php
*** Microarray
:PROPERTIES:
:ID:       294edcbb-f306-4f94-9eba-e8a1c7c512f1
:END:
- cite:lee2004
- cite:nguyen2002
*** Allele-specific expression
:PROPERTIES:
:ID:       ed7d8d19-8da3-4c51-b082-730d1cf1d3ec
:END:

****** Applications for bulk RNA-seq
:PROPERTIES:
:ID:       66741e99-6a43-4ef0-a0f3-12c4bdaaebb0
:END:

:PROPERTIES:
:ID:       a9e9bbf7-8012-456a-9b45-0c86aacd84c7
:END:
- [[*cell enumeration through bulk RNA-seq][cell enumeration through bulk RNA-seq]]
- deconvolution
  - http://labs.pathology.jhu.edu/halushka/science/
  - tools
    - ISOpure
      - https://link.springer.com/article/10.1186/s12859-015-0597-x#Sec1
  - cite:newman2015
- Clinical applications of RNA-seq
  - cite:byron2016translating
  - cite:Kalari_2014
  - cite:Van_Keuren_Jensen_2014
******* eQTL (expression quantitative trait loci) mapping
:PROPERTIES:
:ID:       2CA85A6B-7B1E-4A38-96AA-717847A59168
:END:
- identifies genetic variants that explain variation in gene expression
- w/ RNA-seq via splAdder
  - https://github.com/ratschlab/spladder
  - https://academic.oup.com/bioinformatics/article/32/12/1840/1744052?login=true
*** Spatial transcriptomics
:PROPERTIES:
:ID:       053daf71-9f9f-4b5e-b400-222c4d24afea
:END:
https://www.biorxiv.org/content/10.1101/2023.12.22.572554v1.full
- scopus
- https://www.nature.com/articles/s41592-019-0548-y
- https://www.nature.com/articles/s41586-022-05202-1
- Platforms
  | Platform   | Areas per slide | Barcodes per area | Detection                     |
  |------------+-----------------+-------------------+-------------------------------|
  | 10x Visium |               4 |              5000 | in situ capture, seq by synth |
  | GeoMx      |                 |                   | in situ hyb w/ oligo detect   |
  | ish        |                 |                   | in sity hyb                   |
  - Visium
    - https://mail.google.com/mail/u/0/#inbox/FMfcgxwLtQMfThxFrGwWlHQFgpTTDrJJ
    - https://mail.google.com/mail/u/0/#inbox/FMfcgxwLtGthWWRrdltpWVLTXhQKnZDs https://support.10xgenomics.com/spatial-gene-expression/sequencing/doc/specifications-sequencing-requirements-for-visium-spatial-gene-expression
  - GeoMx
    - https://www.scopus.com/results/results.uri?sort=plf-f&src=s&st1=geomx&sid=cceac18bcbdba4f776ce2a6635ba19bc&sot=b&sdt=b&sl=20&s=TITLE-ABS-KEY%28geomx%29&origin=searchbasic&editSaveSearch=&yearFrom=Before+1960&yearTo=Present
      - Only covers protein- cite:van2019
      - cite:nielsen2020
    - https://mail.google.com/mail/u/0/#inbox/FMfcgzGmvffBFCTQTwfXgXkvZNcgdlrN
  - Ish https://science.sciencemag.org/content/360/6389/660
- Reference
  - https://www.scopus.com/results/results.uri?cc=10&sort=cp-f&src=s&st1=%22spatial+transcriptomics%22&nlo=&nlr=&nls=&sid=f8602d6245cf86196aeef7c88b5ab0dd&sot=b&sdt=cl&cluster=scosubtype%2c%22re%22%2ct&sl=40&s=TITLE-ABS-KEY%28%22spatial+transcriptomics%22%29&ss=cp-f&ps=r-f&editSaveSearch=&origin=resultslist&zone=resultslist
    - https://www.sciencedirect.com/science/article/pii/S0958166916302397?via%3Dihub
    - Proof of concept https://www.science.org/doi/10.1126/science.aaf2403
  - [[file:~/repos/org/career.org::*Zander spatial transcriptomics zoom][Zander spatial transcriptomics zoom]]



 MERFISH2 or SeqFISH3 to averages across cells spanning dozens of micrometers in spot-based protocols such as spatial transcriptomics4. Methods also vary in the number of molecular features acquired: from tens in fluorescence in situ hybridization (FISH), cyclic immunofluorescence (CyCIF)5 and imaging mass cytometry (IMC)6 to hundreds or thousands in specialized probe-based spatial transcriptomics methods (MERFISH2 or SeqFISH3) or imaging mass spectrometry7, and tens of thousands in spot-based spatial transcriptomics, such as Slide-Seq8,9, Visium10, DBiT-seq11 and high-definition spatial transcriptomics (HDST)12. Suc


| Platform    | Description                 | ffpe | multiplex | resolution | species |
|-------------+-----------------------------+------+-----------+------------+---------|
| MERFISH2    | DNA probe amp               | no?  |           |            |         |
| SeqFISH     |                             |      |           |            |         |
| IMC         |                             |      |           |            |         |
| Slide-Seq2  |                             |      |           |            |         |
| Visium ffpe | probe                       | yes  |           |            |         |
| Visium      | onslide cDNA barcode to seq | no   |           |            |         |
| DBiT-seq    |                             |      |           |            |         |
| HDST        |                             |      |           |            |         |
| RNAscope    | DNA probe amp               |      |           |            |         |
| DSP         | Probe cleave to array       | yes  |           |            |         |
*** Expression flux
:PROPERTIES:
:ID:       bd5d8a34-80c8-402f-97c4-273a577629ca
:END:
- https://journals.physiology.org/doi/full/10.1152/physiolgenomics.00128.2018
*** Trajectories
:PROPERTIES:
:ID:       056d6cc2-4649-4818-a891-71ae3343f3d1
:END:
[[https://www.cell.com/cell-reports/pdf/S2211-1247(20)30367-3.pdf]]
*** Co-expression / pathway enrichment
:PROPERTIES:
:ID:       91d46e60-eec3-48c7-97f1-b53eb1559b3f
:END:
- https://www.science.org/doi/10.1126/sciadv.ade2812

https://cran.r-project.org/web/packages/pathfindR/vignettes/intro_vignette.html
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9371296/
https://www.reddit.com/r/bioinformatics/comments/1b37wuh/optimize_gsea_analyses/
https://uclouvain-cbio.github.io/WSBIM2122/sec-gsea.html
[[https://www.reddit.com/r/bioinformatics/comments/13vk8r5/comparing_gsea_results_between_different_species/][clusterprofiler term to gene]]
- cite:reimand2019
- https://old.reddit.com/r/bioinformatics/comments/r7hzoq/annual_ingenuity_subscription_sticker_shock_and/
- https://www.sciencedirect.com/science/article/pii/S2666166721004755
- https://hbctraining.github.io/Training-modules/DGE-functional-analysis/lessons/02_functional_analysis.html

- cite:sanchez-vega2018
https://www.sciencedirect.com/science/article/pii/S2666166721004755
https://www.reddit.com/r/bioinformatics/comments/tu61pw/doubt_over_pathway_analysis_approach/
https://yulab-smu.top/biomedical-knowledge-mining-book/index.html
https://girke.bioinformatics.ucr.edu/GEN242/tutorials/rfea/rfea/
https://bioinformaticsbreakdown.com/how-to-gsea/
https://metascape.org/gp/index.html#/main/step1
https://yulab-smu.top/biomedical-knowledge-mining-book/index.html

- https://www.biorxiv.org/content/10.1101/2021.12.19.473368v1
- variancePartition https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-016-1323-z

**** Methods
:PROPERTIES:
:ID:       912231a2-6288-48f4-85a6-4510ee94de80
:END:
- https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-017-1482-6
- [[id:9b222d1c-ab61-450c-8b77-a98ec3213867][gage]]
  - https://genviz.org/module-04-expression/0004/03/01/pathwayAnalysis/
- Aggregate - clusterProfiler
  - cite:yu2012
  - http://bioconductor.org/packages/release/bioc/html/clusterProfiler.html
- iPANDA https://www.nature.com/articles/ncomms13427#Sec10
***** Over-representation / hypergeometric test
:PROPERTIES:
:ID:       4c4862c9-64f5-4028-a4a9-a7ae76c68ba2
:END:
- https://david.ncifcrf.gov/gene2gene.jsp
- https://maayanlab.cloud/Enrichr/
- Fisher exact
  - https://maayanlab.cloud/Enrichr/
- Proposed in cite:boyle2004
***** [[id:D68E1182-1F7B-4160-A6DD-BD2635F8085F][GSEA]] (Gene set enrichment)
:PROPERTIES:
:ID:       b2ca2c6c-d7f0-41a6-88aa-ec51bc77afa8
:END:
***** Methods and associated tools
:PROPERTIES:
:ID:       d11cda99-d3a3-4549-a828-d065c6979c5d
:END:
- [[id:948289d8-3a77-471a-b068-0507285e0bac][wgcna]]
  - wgcna
    - https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/Tutorials/FemaleLiver-02-networkConstr-man.pdf
    - cite:vanDam2018
      #+begin_src bash
    mkdir -p /tmp/wgnca-example
    cd /tmp/wgnca-example/
    wget https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/Tutorials/FemaleLiver-Data.zip
    unzip FemaleLiver-Data.zip
    #+end_src
      #+begin_src R
  setwd("/tmp/wgnca-example")
  source("~/repos/basecamp/bin/setup.R")
  #
  # ez install of wgcna and dependencies for R 3.6
  if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
  BiocManager::install("fastLiquidAssociation")
  #
  library(WGCNA)
  options(stringsAsFactors=F)
  femData=read.csv("./LiverFemale3600.csv")
  datExpr0=as.data.frame(t(femData[,-c(1:8)]))
  head(datExpr0)
  gsg=goodSamplesGenes(datExpr0, verbose=3)
  gsg$allOK
  if (!gsg$allOK)
  {
  # Optionally, print the gene and sample names that were removed:
  if (sum(!gsg$goodGenes)>0)
  printFlush(paste("Removing genes:", paste(names(datExpr0)[!gsg$goodGenes], collapse = ", ")));
  if (sum(!gsg$goodSamples)>0)
  printFlush(paste("Removing samples:", paste(rownames(datExpr0)[!gsg$goodSamples], collapse = ", ")));
  # Remove the offending genes and samples from the data:
  datExpr0 = datExpr0[gsg$goodSamples, gsg$goodGenes]
  }
  #
  sampleTree=hclust(dist(datExpr0),method="average")
  plot(sampleTree)
  #
  # Plot a line to show the cut
  abline(h = 15, col = "red");
  # Determine cluster under the line
  clust = cutreeStatic(sampleTree, cutHeight = 15, minSize = 10)
  table(clust)
  # clust 1 contains the samples we want to keep.
  keepSamples = (clust==1)
  datExpr = datExpr0[keepSamples, ]
  nGenes = ncol(datExpr)
  nSamples = nrow(datExpr)
  #
  traitData = read.csv("ClinicalTraits.csv")
  dim(traitData)
  names(traitData)
  # remove columns that hold information we do not need.
  allTraits = traitData[, -c(31, 16)]
  allTraits = allTraits[, c(2, 11:36) ]
  dim(allTraits)
  names(allTraits)
  # Form a data frame analogous to expression data that will hold the clinical traits.
  femaleSamples = rownames(datExpr)
  traitRows = match(femaleSamples, allTraits$Mice)
  datTraits = allTraits[traitRows, -1]
  rownames(datTraits) = allTraits[traitRows, 1]
  collectGarbage()
  #
  save(datExpr, datTraits, file = "FemaleLiver-01-dataInput.RData")
  #
  enableWGCNAThreads()
  lnames=load(file="FemaleLiver-01-dataInput.RData")
  #
  lnames
    #+end_src
**** Databases
:PROPERTIES:
:ID:       f7eed3be-f505-451c-b191-1a58373fb180
:END:
- [[id:c79bfe84-0f15-474e-8a22-1f210605ed5a][MSigDB]]
- https://academic.oup.com/bioinformatics/article/33/3/414/2875813
- https://metascape.org/gp/index.html#/main/step1
- gene sets for mouse GSEA
  - https://cran.r-project.org/web/packages/msigdbr/vignettes/msigdbr-intro.html
  - recent, high impact papers citing MSigDB and aligning to mm10
    - https://science.sciencemag.org/content/352/6288/aad3018
    - http://europepmc.org/backend/ptpmcrender.fcgi?accid=PMC5924693&blobtype=pdf
    - https://www.cell.com/cell/pdf/S0092-8674(17)31385-5.pdf
    - https://stm.sciencemag.org/content/scitransmed/suppl/2018/01/12/10.424.eaan5488.DC1/aan5488_SM.pdf
**** Applciation Examples
:PROPERTIES:
:ID:       1139b0a7-32a3-4ccf-8bf2-0266f5e9b2e5
:END:
- cite:sanchez-vega2018
**** Gene regulatory network inference
:PROPERTIES:
:ID:       b27fe952-97b2-447b-b97d-08f38d68be87
:END:
- tools
  - GENIE
    - https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0012776
    - https://www.bioconductor.org/packages/devel/bioc/vignettes/GENIE3/inst/doc/GENIE3.html
**** Gene co-expression visualization
:PROPERTIES:
:ID:       fe3baa6c-d28e-48c4-91b1-0539e01a4cb1
:END:
- Need ssGSEA / GSVA for pathway heatmap https://bioconductor.org/packages/release/bioc/vignettes/GSVA/inst/doc/GSVA.html
- [[https://enrichmentmap.readthedocs.io/en/docs-2.2/QuickStartGuide.html][enrichement map]]
  - https://enrichmentmap.readthedocs.io/en/latest/Parameters.html
- https://kiharalab.org/web/navigo/views/goset.php
- http://www.visantnet.org/visantnet.html
- [[https://bioconductor.org/packages/release/bioc/vignettes/pathview/inst/doc/pathview.pdf][Pathview]]
  - https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9653450/
- visualization
  - heatmaps
    - https://www.frontiersin.org/files/Articles/425394/fgene-10-00858-HTML/image_m/fgene-10-00858-g002.jpg
    - https://www.bioconductor.org/packages/release/bioc/vignettes/ReactomePA/inst/doc/ReactomePA.html
  - wire diagrams
  - bar and lolipop for p-value and set size
    - https://www.nature.com/articles/s41598-018-29420-8
    - https://www.frontiersin.org/articles/10.3389/fgene.2019.00858/full
- https://github.com/PathwayCommons/chibe
- https://github.com/Adrienrougny/sbgntikz/
**** Tutorials
:PROPERTIES:
:ID:       1ff81f54-552c-40f6-bc19-10aefecd034f
:END:
- https://bioinformaticsbreakdown.com/how-to-gsea/
- Nature Protocol: cite:reimand2019
- tutorial: https://girke.bioinformatics.ucr.edu/GEN242/tutorials/rfea/rfea/
**** General reference
:PROPERTIES:
:ID:       98048a64-6bd5-4107-ab7e-573b57758f12
:END:
- cite:maleki2020
- [[id:2CA85A6B-7B1E-4A38-96AA-717847A59168][QTL mapping]]
- tutorial on transcriptional regulatory networks from total RNA-seq cite:chouvarine2021
- cite:vandam2018
- https://yulab-smu.top/biomedical-knowledge-mining-book/enrichment-overview.html
- pathway annotation and conservation
  - https://academic.oup.com/nar/article/48/D1/D479/5626534#190997651
- method and tool comparisons
  - cite:amadoz2019
- https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1790-4
- cite:reimand2019
- https://link.springer.com/article/10.1186/s40246-019-0226-2
- https://www.sciencedirect.com/science/article/pii/S0022519314001969?via%3Dihub
- cite:creixell2015
- cite:rahmatallah2016

**** Methods and associated tools
:PROPERTIES:
:ID:       d11cda99-d3a3-4549-a828-d065c6979c5d
:END:
- https://maayanlab.cloud/Enrichr/
- [[id:2CA85A6B-7B1E-4A38-96AA-717847A59168][QTL mapping]]
- iPANDA
  - https://www.nature.com/articles/ncomms13427#Sec10
    - https://github.com/varnivey/ipanda/blob/master/MANUAL.txt
- Aggregate - clusterProfiler
  - http://yulab-smu.top/clusterProfiler-book/index.html
  - http://bioconductor.org/packages/release/bioc/html/clusterProfiler.html
- Gene set enrichment
  - https://david.ncifcrf.gov/gene2gene.jsp
  - https://genviz.org/module-04-expression/0004/03/01/pathwayAnalysis/
  - https://monashbioinformaticsplatform.github.io/RNAseq-DE-analysis-with-R/RNAseq_DE_analysis_with_R.html#gene-set-enrichment
  - http://software.broadinstitute.org/gsea/msigdb/collections.jsp
  - https://monashbioinformaticsplatform.github.io/RNAseq-DE-analysis-with-R/RNAseq_DE_analysis_with_R.html#gene-set-enrichment
  - pathway enrichment
    - - https://bioconductor.org/packages/release/bioc/vignettes/gage/inst/doc/RNA-seqWorkflow.pdf
    - - https://www.nature.com/articles/s41596-018-0103-9.pdf?proof=true19
    - https://bioconductor.org/packages/release/bioc/vignettes/ReactomePA/inst/doc/ReactomePA.html
  - [[id:9b222d1c-ab61-450c-8b77-a98ec3213867][gage]]
  - cite:rahmatallah2014
  - https://academic.oup.com/bioinformatics/article/33/3/414/2875813
  - [[id:D68E1182-1F7B-4160-A6DD-BD2635F8085F][GSEA]]
- Fisher exact
  - https://maayanlab.cloud/Enrichr/
- [[id:948289d8-3a77-471a-b068-0507285e0bac][wgcna]]
  - wgcna
    - https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/Tutorials/FemaleLiver-02-networkConstr-man.pdf
    - cite:vanDam2018
      #+begin_src bash
    mkdir -p /tmp/wgnca-example
    cd /tmp/wgnca-example/
    wget https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/Tutorials/FemaleLiver-Data.zip
    unzip FemaleLiver-Data.zip
    #+end_src
      #+begin_src R
  setwd("/tmp/wgnca-example")
  source("~/repos/basecamp/bin/setup.R")
  #
  # ez install of wgcna and dependencies for R 3.6
  if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
  BiocManager::install("fastLiquidAssociation")
  #
  library(WGCNA)
  options(stringsAsFactors=F)
  femData=read.csv("./LiverFemale3600.csv")
  datExpr0=as.data.frame(t(femData[,-c(1:8)]))
  head(datExpr0)
  gsg=goodSamplesGenes(datExpr0, verbose=3)
  gsg$allOK
  if (!gsg$allOK)
  {
  # Optionally, print the gene and sample names that were removed:
  if (sum(!gsg$goodGenes)>0)
  printFlush(paste("Removing genes:", paste(names(datExpr0)[!gsg$goodGenes], collapse = ", ")));
  if (sum(!gsg$goodSamples)>0)
  printFlush(paste("Removing samples:", paste(rownames(datExpr0)[!gsg$goodSamples], collapse = ", ")));
  # Remove the offending genes and samples from the data:
  datExpr0 = datExpr0[gsg$goodSamples, gsg$goodGenes]
  }
  #
  sampleTree=hclust(dist(datExpr0),method="average")
  plot(sampleTree)
  #
  # Plot a line to show the cut
  abline(h = 15, col = "red");
  # Determine cluster under the line
  clust = cutreeStatic(sampleTree, cutHeight = 15, minSize = 10)
  table(clust)
  # clust 1 contains the samples we want to keep.
  keepSamples = (clust==1)
  datExpr = datExpr0[keepSamples, ]
  nGenes = ncol(datExpr)
  nSamples = nrow(datExpr)
  #
  traitData = read.csv("ClinicalTraits.csv")
  dim(traitData)
  names(traitData)
  # remove columns that hold information we do not need.
  allTraits = traitData[, -c(31, 16)]
  allTraits = allTraits[, c(2, 11:36) ]
  dim(allTraits)
  names(allTraits)
  # Form a data frame analogous to expression data that will hold the clinical traits.
  femaleSamples = rownames(datExpr)
  traitRows = match(femaleSamples, allTraits$Mice)
  datTraits = allTraits[traitRows, -1]
  rownames(datTraits) = allTraits[traitRows, 1]
  collectGarbage()
  #
  save(datExpr, datTraits, file = "FemaleLiver-01-dataInput.RData")
  #
  enableWGCNAThreads()
  lnames=load(file="FemaleLiver-01-dataInput.RData")
  #
  lnames
    #+end_src
**** General reference
:PROPERTIES:
:ID:       98048a64-6bd5-4107-ab7e-573b57758f12
:END:
- https://yulab-smu.top/biomedical-knowledge-mining-book/enrichment-overview.html
- pathway annotation and conservation
  - https://academic.oup.com/nar/article/48/D1/D479/5626534#190997651
- method and tool comparisons
  - cite:amadoz2019
- https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1790-4
- cite:reimand2019
- https://link.springer.com/article/10.1186/s40246-019-0226-2
- https://www.sciencedirect.com/science/article/pii/S0022519314001969?via%3Dihub
- cite:creixell2015
- cite:rahmatallah2016
*** Gene expression dynamics
:PROPERTIES:
:ID:       86071a18-2c82-4ca0-93a5-ef6f2dd9f0a4
:END:
- [[https://www.scopus.com/results/citedbyresults.uri?sort=cp-f&cite=2-s2.0-0344464733&src=s&nlo=&nlr=&nls=&imp=t&sid=1aade77c6b42db34d11646993e38a758&sot=cite&sdt=a&sl=0&origin=resultslist&offset=1&txGid=3af857e8bcb60fab910d99917b405130][scopus citing GEDI]]
**** Time course
:PROPERTIES:
:ID:       43f1772e-f8ce-4f6b-a4bd-d35d381b89c9
:END:
  - cite:spies2017
  - https://support.bioconductor.org/p/65676/#66860
  - https://www.biostars.org/p/138988/
  - https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#group-specific-condition-effects-individuals-nested-within-groups
  - https://hbctraining.github.io/DGE_workshop/lessons/08_DGE_LRT.html
  - http://seqanswers.com/forums/showthread.php?t=64039
  - https://www.sciencedirect.com/science/article/pii/S2001037015000392
** Repository setup and administration
*** Bash preamble
*** prebuilt salmon refs
http://refgenomes.databio.org/v3/genomes/splash/2230c535660fb4774114bfa966a62f823fdb6d21acf138d4

*** Setup
- get test data
  #+begin_src bash
repo=~/repos/rna-seq
# get salmon files

cp ~/card/inputs/Rentschler_s4630_MGI0042/1-ir.AACATCTCGA-TATTCGCCAG/1-ir.AACATCTCGA-TATTCGCCAG.quant.sf ${repo}/test/inputs/exp1.quant.sf

cp ~/card/inputs/Rentschler_s4630_MGI0042/2-ir.GTAACGTCAC-ACCAACTAAG/2-ir.GTAACGTCAC-ACCAACTAAG.quant.sf ${repo}/test/inputs/exp2.quant.sf

cp ~/card/inputs/Rentschler_s4630_MGI0042/3-ir.GCTAACCGTG-TCACCATAAG/3-ir.GCTAACCGTG-TCACCATAAG.quant.sf ${repo}/test/inputs/expsample3.quant.sf

cp ~/card/inputs/Rentschler_s4630_MGI0042/10-ct.TCATCCGTGA-TTAGGAGGAA/10-ct.TCATCCGTGA-TTAGGAGGAA.quant.sf ${repo}/test/inputs/ctrl.quant.sf

cp ~/card/inputs/Rentschler_s4630_MGI0042/9-ct.TCCAGAATGT-TGGTCCAATT/9-ct.TCCAGAATGT-TGGTCCAATT.quant.sf ${repo}/test/inputs/ctrl2.quant.sf

cp ~/card/inputs/Rentschler_s4630_MGI0042/8-ct.CCGAACATGT-ACCTCATTGA/8-ct.CCGAACATGT-ACCTCATTGA.quant.sf ${repo}/test/inputs/normal.quant.sf
#+end_src
- file:test/inputs/libraries.tsv
  | library | basename            | participant | run  | group |
  |---------+---------------------+-------------+------+-------|
  | lib001  | ctrl.quant.sf       | subj001     | seq1 | ctrl  |
  | lib002  | ctrl2.quant.sf      | subj002     | seq2 | ctrl  |
  | lib003  | normal.quant.sf     | subj003     | seq1 | ctrl  |
  | lib004  | exp1.quant.sf       | subj004     | seq2 | exp   |
  | lib005  | exp2.quant.sf       | subj005     | seq1 | exp   |
  | lib006  | expsample3.quant.sf | subj006     | seq2 | exp   |
** Bulk RNA-seq Exploratory Data Analysis                               :smk:
:PROPERTIES:
:header-args:snakemake: :tangle ./workflow/rna_seq_eda.smk
:END:
*** Make txi counts from salmon files                              :smk_rule:
- Snakemake
  #+begin_src snakemake
rule make_salmon_txi:
    input: expand(salmon_dir + "/{library}.sf", library = RNA_LIBS),
    log: logdir + "/{experiment}_make_salmon_txi.log",
    output: rna_dir + "/{experiment}_txi.rdata",
    params:
        script = rna_script_dir + "/make_salmon_txi.R",
        txdb = txdb,
    shell:
        """
        Rscript {params.script} \
        "{input}" \
        {output} \
        {params.txdb} \
        > {log} 2>&1
        """
#+end_src
- [[file:./workflow/scripts/make_salmon_txi.R][Rscript]]
  #+begin_src R :noweb yes :tangle ./scripts/make_salmon_txi.R
#!/usr/bin/env Rscript


args = commandArgs(trailingOnly = TRUE)
in_salmon_str = args[1]
out_txi = args[2]
in_txdb = args[3]

# Load libraries
library(paste(in_txdb), character.only=T)
txdb = get(in_txdb)
library(tximport)

# Make salmon file list
in_salmon_vec = unlist(strsplit(in_salmon_str, " "))
names(in_salmon_vec) = substr(gsub("^.*lib", "lib", in_salmon_vec), 1, 6)

# Make gene annotation
k = keys(txdb, keytype = "TXNAME")
tx2gene = AnnotationDbi::select(txdb, k, "GENEID", "TXNAME")

# Make txi object
txi = tximport(in_salmon_vec, type = "salmon", tx2gene = tx2gene)

# Save txi object
save(txi, file = out_txi)
#+end_src
*** INPROCESS All sample EDA                                       :smk_rule:
- Snakemake
  #+begin_src snakemake
rule all_rna_eda:
    container: "/home/jeszyman/sing_containers/atac.1.1.0.sif",
    input: analysis + "/{experiment}_txi.rdata",
    log: logdir + "/{experiment}_rna_eda.log",
    output:
        pca = results + "/figures/{experiment}_pca.pdf",
        rdata = analysis + "/{experiment}_eda.rdata",
    params:
        factor_str = factor_str,
        library_tsv = library_tsv,
        script = rna_scriptdir + "/all_rna_eda.R",
    shell:
        """
        Rscript {params.script} \
        {input} \
        {output.pca} \
        {output.rdata} \
        "{params.factor_str}" \
        {params.library_tsv} \
        > {log} 2>&1
        """
#+end_src
- Rscript
  #+begin_src R :tangle ./scripts/all_rna_eda.R
#!/usr/bin/env Rscript
args = commandArgs(trailingOnly = TRUE)
txi_rdata = args[1]
pca_plot_pdf = args[2]
out_rdata = args[3]
factor_str = args[4]
library_tsv = args[5]

## txi_rdata = "test/analysis/all_txi.rdata"
## library_tsv = "test/inputs/libraries.tsv"
## factor_str = "run group"
## out_rdata = "test/analysis/eda.rdata"
## pca_plot_pdf = "test/results/all_pca.pdf"

library(cowplot)

library(ggrepel)
library(tidyverse)

load(txi_rdata)

library(edgeR)

counts = txi$counts

# Obtaining per-observation scaling factors for length, adjusted to avoid
# (see https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html#edgeR)
norm_mat = txi$length
norm_mat = norm_mat/exp(rowMeans(log(norm_mat)))
norm_counts = counts/norm_mat

# Get effective library sizes from scaled counts
eff_lib = calcNormFactors(norm_counts) * colSums(norm_counts)
norm_mat = sweep(norm_mat, 2, eff_lib, "*")
norm_mat = log(norm_mat)

# Creating a DGEList object for use in edgeR.
y = DGEList(counts)
y = scaleOffset(y, norm_mat)

libraries = read_tsv(library_tsv)
factor_vec = unlist(strsplit(factor_str, " "))

formula = as.formula(paste("~ ", paste(factor_vec, collapse = "+")))
formula
design = model.matrix(formula, libraries)
design

# filtering using the design information
## design <- model.matrix(~condition, data = sampleTable)
keep <- filterByExpr(y, design)
y <- y[keep, ]

logCPM <- cpm(y, prior.count=2, log=TRUE, offset = y$offset)

pca = prcomp(t(logCPM))

make_pca_plots = function(in_pca, full_libs){
  pve_pc1=round(100*summary(in_pca)$importance[2,1])
  pve_pc2=round(100*summary(in_pca)$importance[2,2])
  pca_plot = as.data.frame(in_pca$x) %>%
    rownames_to_column(var = "library") %>%
    left_join(libraries, by = "library") %>%
    ggplot(., aes(x = PC1, y = PC2, color = get(tail(factor_vec, n= 1)), label = library)) +
    geom_point(size = 4) +
    geom_text_repel() +
    xlab(paste("PC1, ", pve_pc1, "% variance explained", sep ="")) +
    ylab(paste("PC2, ", pve_pc2, "% variance explained", sep ="")) +
    scale_color_discrete(name = paste0(tail(factor_vec, n=1))) +
    theme_cowplot() +
    theme(legend.position = "bottom")
  return(pca_plot)
}

pca_plot = make_pca_plots(pca, libraries)
# Note this resembles plotMDS(y, gene.selection = "common")

save(design, formula, logCPM, pca, pca_plot, y, file = out_rdata)

save_plot(pca_plot, file = pca_plot_pdf)
#+end_src
*** Dev
:PROPERTIES:
:header-args:snakemake: :tangle no
:END:
*** Ideas
:PROPERTIES:
:header-args:snakemake: :tangle no
:END:
** Snakemake conda environment yaml
:PROPERTIES:
:header-args:bash: :tangle ./config/rna_env.yaml
:END:
#+begin_src bash

name: rna
channels:
  - bioconda
  - conda-forge
dependencies:
  - bioconductor-deseq2
  - bioconductor-txdb.hsapiens.ucsc.hg38.knowngene
  - bioconductor-tximport
  - r-tidyverse

#+end_src
** Snakemake configuration YAML
#+begin_src bash :tangle ./config/int_test.yaml
experiment_id: inttest_human
txdb: "TxDb.Mmusculus.UCSC.mm10.ensGene",
threads: 4
rna_container: "/home/jeszyman/sing_containers/rna.1.2.0.sif"
datadir: "test"
rna_repo: "/home/jeszyman/repos/rna-seq"
factor_str: "run group"
#+end_src
** [[file:workflow/int_test.smk][Integration testing]]                  :smk:
:PROPERTIES:
:header-args:snakemake: :tangle ./workflow/int_test.smk
:END:
*** Preamble
#+begin_src snakemake
#########1#########2#########3#########4#########5#########6#########7#########8
###                                                                          ###
###               Integration Testing Snakefile for RNA-seq                  ###
###                                                                          ###
#########1#########2#########3#########4#########5#########6#########7#########8

##################################
###   Load Required Packages   ###
##################################

import numpy as np
import os
import pandas as pd
import re

#+end_src
*** Variable naming
#+begin_src snakemake

###########################
###   Variable Naming   ###
###########################

# Names directly from configuration YAML
threads = config['threads']

# Names build from configuration parameter base
rna_script_dir = config['rna_repo'] + "/scripts"
#+end_src
*** Functions
#+begin_src snakemake
RNA_LIBS = ["lib001", "lib002", "lib003", "lib004"]
#+end_src
*** All rule
#+begin_src snakemake
rule all:
    input:
        analysis + "/{experiment}_txi.rdata",
rule dumbtest:
    output: "/tmp/test2.tsv",
    params:
        script = rna_script_dir + "/dumbtest.R"
    shell:
        """
        Rscript {params.script} \
        {output}
        """
#+end_src
- [[file:./scripts/dumbtest.R][Rscript]]
  #+begin_src R :tangle ./scripts/dumbtest.R
#!/usr/bin/env Rscript

############
###      ###
############

# For unit testing


# Command line arguements
args = commandArgs(trailingOnly = TRUE)
out= args[1]

# Load required packages
library(tidyverse)

test = data.frame(top=c(1,2,3),
                  bottom=c('a','b','c'))

test2 = as_tibble(test)

write_tsv(test2, file = out)
#+end_src
*** Benchmark aggregation
*** Includes statements
*** Bulk RNA-seq Exploratory Data Analysis Integration Testing          :smk:
:PROPERTIES:
:header-args:snakemake: :tangle ./workflow/rna_seq_eda_int.smk
:END:
**** Preamble
#+begin_src snakemake
#print("Integration testing snakefile for bulk RNA-seq\n")

# Import common packages
import pandas as pd
import re
import numpy as np

#+end_src
**** Variable naming
#+begin_src snakemake
datadir = config["datadir"]
inputs=datadir + "/inputs"
analysis = datadir + "/analysis"
salmon = analysis + "/salmon"
results = datadir + "/results"
factor_str= config["factor_str"]
rna_container = config["rna_container"]
logdir = config["datadir"] + "/logs"

rna_repo = config["rna_repo"]
rna_scriptdir = rna_repo + "/scripts"
library_tsv=inputs + "/libraries.tsv"
#+end_src
**** Functions, miscellaneous
#+begin_src snakemake
rna_libraries = pd.read_table(inputs + "/libraries.tsv")
rna_libraries["path"]= inputs + "/" + rna_libraries["basename"]

# Needs full path to work (no tilda)
readable = []
for x in rna_libraries.path:
    readable.append(os.access(x, os.R_OK))
rna_libraries['readable']=readable

rna_libraries = rna_libraries[rna_libraries.readable == True]

rna_library_indict = rna_libraries["library"].tolist()
rna_file_indict = rna_libraries["path"].tolist()
rna_lib_dict = dict(zip(rna_library_indict, rna_file_indict))

BULK_RNA_LIBS = list(rna_lib_dict.keys())

#+end_src
**** All rule
#+begin_src snakemake
rule all:
    input:
        expand(salmon + "/{library}.quant.sf", library = BULK_RNA_LIBS),
        expand(analysis + "/{experiment}_txi.rdata", experiment = "all"),
        results + "/figures/all_pca.pdf",
        analysis + "/all_eda.rdata",
#+end_src
**** Symlink inputs                                                :smk_rule:
- Snakemake
  #+begin_src snakemake
rule symlink_salmon:
    container: rna_container,
    input: lambda wildcards: rna_lib_dict[wildcards.library],
    log: logdir + "/{library}_symlink_salmon.log",
    output: salmon + "/{library}.quant.sf",
    params:
        script = rna_scriptdir + "/symlink_salmon.sh"
    shell:
        """
        {params.script} {input} {output} &> {log}
        """
#+end_src
- Shell
  #+begin_src bash :tangle ./scripts/symlink_salmon.sh
#!/usr/bin/env bash
set -o errexit   # abort on nonzero exitstatus
set -o nounset   # abort on unbound variable
set -o pipefail  # don't hide errors within pipes

variables(){
   in_salmon="${1}"
   out_sym="${2}"
}

main(){
    variables $@
    ln --force --relative --symbolic $in_salmon $out_sym
}

main "$@"

#+end_src
**** Include statements
#+begin_src snakemake
include: rna_repo + "/workflow/rna_seq_eda.smk"
#+end_src

*** Post-QC RNA-seq Differential Expression                             :smk:
:PROPERTIES:
:header-args:snakemake: :tangle ./workflow/rnaseq_de.smk
:END:
**** Preamble
#+begin_src snakemake
print("Integration testing snakefile for Post-QC RNA-seq Differential Expression\n")

# Import common packages
import pandas as pd
import re
import numpy as np

#+end_src
**** Variable naming
#+begin_src snakemake

#+end_src
**** Functions, miscellaneous
#+begin_src snakemake

#+end_src
**** All rule
#+begin_src snakemake
rule all:
    input:
        design
        tmm
        ebayes
        dds
#+end_src
**** Symlink inputs
#+begin_src snakemake
rule symlink_rnaseq_de_inputs:
    input:

#+end_src
**** Include statements
#+begin_src snakemake
#include: " <INCLUDE FILE LOCATION (VIA CONFIG PARAM)>"
#+end_src
** Repository setup and administration
*** DONE [[id:1a8d792b-9621-42a8-80c9-399b6065630a][Bioinformatics pipeline initial build]]
CLOSED: [2022-10-26 Wed 13:48]
** README
:PROPERTIES:
:export_file_name: ./readme.md
:export_file_options: toc:nil
:END:
*** Pipeline Directed Acyclic Graphs
*** Prerequisites for local integration testing
- See configuation file (./config/int_test.yaml) for required parameters and versioned singularity container
*** Changelog
- [2022-10-28 Fri] Stable 1.0.1 libraries_tsv naming abstracted
- [2022-10-28 Fri] Stable 1.0.0 Basic edgeR run

** Development                                                          :dev:
*** Version 1.0.0
**** TODO [[id:f6717c79-64ce-4b16-b455-649df2ba20fd][Project stable version update]]
*** Ideas                                                             :ideas:
- cibersort
** General reference
:PROPERTIES:
:ID:       70491293-d832-4947-b1bd-a0f4ae066c17
:END:
- transfer learning in gene expression https://www.sciencedirect.com/science/article/pii/S240547121930119X
- https://www.cell.com/fulltext/S0092-8674%2812%2901226-3
- http://gepia.cancer-pku.cn/
- Regulation of gene expression https://www.reddit.com/r/bioinformatics/comments/sjp16n/database_of_differential_gene_expression_based_on/

- my RNA-seq work
  - [[file:~/org/career.org::*high-grade head and neck characterization][high-grade head and neck characterization]]
  - [[id:e7a8362d-f4cf-45ec-a3f8-91b0ffbdc500][ffpe rna-seq project]]
  - [[file:~/data/career/pathology/molecular-pathology/ffpe-rna/ffpe-rnaseq.org]]
- [[https://www.nature.com/subjects/rna-sequencing/nmeth][Nature Methods RNA sequencing]]
- tool comparisons cite:sahraeian2017
- data
  - public RNA sequencing data
    - [[https://webshare.bioinf.unc.edu/public/mRNAseq_TCGA/UNC_mRNAseq_summary.pdf][TCGA RSEM]]
      - https://support.bioconductor.org/p/131066/
      - http://zyxue.github.io/2017/06/02/understanding-TCGA-mRNA-Level3-analysis-results-files-from-firebrose.html
      - https://bsbludwig.com/post/94066296740/what-do-tcgas-rnaseq-files-actually-show
      - http://seqanswers.com/forums/showthread.php?t=42911
      - https://support.bioconductor.org/p/65890/
  - see [[http://www.cbil.upenn.edu/BEERS/][BEERS]] for generating fake RNA seq data
- cite:korpelainen2015
  - [[http://rnaseq-book.blogspot.fi/][Korpelainen,15 web companion]]
- [[http://rnaseq.uoregon.edu/][RNAcyclopedia]]
- https://en.wikipedia.org/wiki/List_of_RNA-Seq_bioinformatics_tools
- [[id:386d899a-9208-4d5c-93d8-3750b22d9748][RNA methods]]
*** Ideas
:PROPERTIES:
:ID:       B5E9166C-CB80-4578-BD90-1E694B46F192
:END:
- http://gepia.cancer-pku.cn/
[[http://bioinf.wehi.edu.au/marray/jsm2005/lab4/lab4.html][A limma workflow to try]]
[[http://combine-australia.github.io/RNAseq-R/06-rnaseq-day1.html][another workflow example]]

see Love,16 section "Rich visualization and reporting of results"

**** Pipeline Comparisons
:PROPERTIES:
:ID:       7B9AD35F-1511-4CB2-BB5A-0989EFEAADF3
:END:
>>>>>>> e60be068dcbc8c18477fc28e00ee3b742dcf0110

[[file:~/Box%20Sync/library/Fonseca,14%20RNA%20seq%20gene%20profiling%20a%20systematic%20empirical%20comparison.PDF][Fonseca,14 RNA seq gene profiling a systematic empirical comparison]]
[[file:~/Box%20Sync/library/Sigurgeirsson,14%20Analysis%20of%20stranded%20information%20using%20an%20automated%20procedure%20for%20strand%20specific%20RNA%20sequencing.pdf][Sigurgeirsson,14 Analysis of stranded information using an automated procedure for strand specific RNA

*** Reference                                                           :ref:
- https://github.com/zhxiaokang/RASflow/blob/master/scripts/dea_genome.R
- https://bitbucket.org/cfce/viper/src/master/modules/scripts/DEseq.R
